## Overview

This notebook will show you how to create and query a table or DataFrame that you uploaded to DBFS. [DBFS](https://docs.databricks.com/user-guide/dbfs-databricks-file-system.html) is a Databricks File System that allows you to store data for querying inside of Databricks. This notebook assumes that you have a file already inside of DBFS that you would like to read from.

This notebook is written in **Python** so the default cell type is Python. However, you can use different languages by using the `%LANGUAGE` syntax. Python, Scala, SQL, and R are all supported.

In [2]:
# File location and type
file_location = "/FileStore/tables/train.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration id2875421 2 2016-03-14 17:24:55 2016-03-14 17:32:30 1 -73.982154846191406 40.767936706542969 -73.964630126953125 40.765602111816406 N 455 id2377394 1 2016-06-12 00:43:35 2016-06-12 00:54:38 1 -73.980415344238281 40.738563537597656 -73.999481201171875 40.731151580810547 N 663 id3858529 2 2016-01-19 11:35:24 2016-01-19 12:10:48 1 -73.979026794433594 40.763938903808594 -74.005332946777344 40.710086822509766 N 2124 id3504673 2 2016-04-06 19:32:31 2016-04-06 19:39:40 1 -74.010040283203125 40.719970703125 -74.01226806640625 40.706718444824219 N 429 id2181028 2 2016-03-26 13:30:55 2016-03-26 13:38:10 1 -73.973052978515625 40.793209075927734 -73.972923278808594 40.782520294189453 N 435 id0801584 2 2016-01-30 22:01:40 2016-01-30 22:09:03 6 -73.982856750488281 40.742195129394531 -73.992080688476562 40.749183654785156 N 443 id1813257 1 2016-06-17 22:34:59 2016-06-17 22:40:40 4 -73.969017028808594 40.757839202880859 -73.957405090332031 40.765895843505859 N 341 id1324603 2 2016-05-21 07:54:58 2016-05-21 08:20:49 1 -73.969276428222656 40.797779083251953 -73.922470092773438 40.76055908203125 N 1551 id1301050 1 2016-05-27 23:12:23 2016-05-27 23:16:38 1 -73.999481201171875 40.738399505615234 -73.985786437988281 40.732814788818359 N 255 id0012891 2 2016-03-10 21:45:01 2016-03-10 22:05:26 1 -73.981048583984375 40.744338989257813 -73.972999572753906 40.789989471435547 N 1225 id1436371 2 2016-05-10 22:08:41 2016-05-10 22:29:55 1 -73.982650756835938 40.763839721679688 -74.002227783203125 40.732990264892578 N 1274 id1299289 2 2016-05-15 11:16:11 2016-05-15 11:34:59 4 -73.991531372070313 40.749439239501953 -73.95654296875 40.7706298828125 N 1128 id1187965 2 2016-02-19 09:52:46 2016-02-19 10:11:20 2 -73.962982177734375 40.756679534912109 -73.984405517578125 40.760719299316406 N 1114 id0799785 2 2016-06-01 20:58:29 2016-06-01 21:02:49 1 -73.956306457519531 40.767940521240234 -73.966110229492188 40.76300048828125 N 260 id2900608 2 2016-05-27 00:43:36 2016-05-27 01:07:10 1 -73.992195129394531 40.727226257324219 -73.974655151367188 40.783069610595703 N 1414 id3319787 1 2016-05-16 15:29:02 2016-05-16 15:32:33 1 -73.955513000488281 40.768592834472656 -73.948760986328125 40.77154541015625 N 211 id3379579 2 2016-04-11 17:29:50 2016-04-11 18:08:26 1 -73.991165161132813 40.755561828613281 -73.999290466308594 40.725353240966797 N 2316 id1154431 1 2016-04-14 08:48:26 2016-04-14 09:00:37 1 -73.994255065917969 40.745803833007813 -73.999656677246094 40.723342895507813 N 731 id3552682 1 2016-06-27 09:55:13 2016-06-27 10:17:10 1 -74.003982543945313 40.7130126953125 -73.979194641113281 40.749923706054688 N 1317 id3390316 2 2016-06-05 13:47:23 2016-06-05 13:51:34 1 -73.98388671875 40.738197326660156 -73.991203308105469 40.727870941162109 N 251 id2070428 1 2016-02-28 02:23:02 2016-02-28 02:31:08 1 -73.980369567871094 40.742420196533203 -73.962852478027344 40.760635375976563 N 486 id0809232 2 2016-04-01 12:12:25 2016-04-01 12:23:17 1 -73.979537963867188 40.753360748291016 -73.963996887207031 40.763458251953125 N 652 id2352683 1 2016-04-09 03:34:27 2016-04-09 03:41:30 1 -73.995864868164063 40.758811950683594 -73.993324279785156 40.740322113037109 N 423 id1603037 1 2016-06-25 10:36:26 2016-06-25 10:55:49 1 -73.993553161621094 40.747173309326172 -74.006141662597656 40.704383850097656 N 1163 id3321406 2 2016-06-03 08:15:05 2016-06-03 08:56:30 1 -73.955230712890625 40.777133941650391 -73.788749694824219 40.641471862792969 N 2485 id0129640 2 2016-02-14 13:27:56 2016-02-14 13:49:19 1 -73.956581115722656 40.771358489990234 -73.974967956542969 40.732791900634766 N 1283 id3587298 1 2016-02-27 21:56:01 2016-02-27 22:14:51 1 -73.9837646484375 40.749874114990234 -73.958831787109375 40.800960540771484 N 1130 id2104175 1 2016-06-20 23:07:16 2016-06-20 23:18:50 1 -73.95843505859375 40.713191986083984 -73.949539184570312 40.6802

In [3]:
# Create a view or table

temp_table_name = "train_csv"

df.createOrReplaceTempView(temp_table_name)

In [4]:
%sql

/* Query the created temp table in a SQL cell */

select * from `train_csv`

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration id2875421 2 2016-03-14 17:24:55 2016-03-14 17:32:30 1 -73.982154846191406 40.767936706542969 -73.964630126953125 40.765602111816406 N 455 id2377394 1 2016-06-12 00:43:35 2016-06-12 00:54:38 1 -73.980415344238281 40.738563537597656 -73.999481201171875 40.731151580810547 N 663 id3858529 2 2016-01-19 11:35:24 2016-01-19 12:10:48 1 -73.979026794433594 40.763938903808594 -74.005332946777344 40.710086822509766 N 2124 id3504673 2 2016-04-06 19:32:31 2016-04-06 19:39:40 1 -74.010040283203125 40.719970703125 -74.01226806640625 40.706718444824219 N 429 id2181028 2 2016-03-26 13:30:55 2016-03-26 13:38:10 1 -73.973052978515625 40.793209075927734 -73.972923278808594 40.782520294189453 N 435 id0801584 2 2016-01-30 22:01:40 2016-01-30 22:09:03 6 -73.982856750488281 40.742195129394531 -73.992080688476562 40.749183654785156 N 443 id1813257 1 2016-06-17 22:34:59 2016-06-17 22:40:40 4 -73.969017028808594 40.757839202880859 -73.957405090332031 40.765895843505859 N 341 id1324603 2 2016-05-21 07:54:58 2016-05-21 08:20:49 1 -73.969276428222656 40.797779083251953 -73.922470092773438 40.76055908203125 N 1551 id1301050 1 2016-05-27 23:12:23 2016-05-27 23:16:38 1 -73.999481201171875 40.738399505615234 -73.985786437988281 40.732814788818359 N 255 id0012891 2 2016-03-10 21:45:01 2016-03-10 22:05:26 1 -73.981048583984375 40.744338989257813 -73.972999572753906 40.789989471435547 N 1225 id1436371 2 2016-05-10 22:08:41 2016-05-10 22:29:55 1 -73.982650756835938 40.763839721679688 -74.002227783203125 40.732990264892578 N 1274 id1299289 2 2016-05-15 11:16:11 2016-05-15 11:34:59 4 -73.991531372070313 40.749439239501953 -73.95654296875 40.7706298828125 N 1128 id1187965 2 2016-02-19 09:52:46 2016-02-19 10:11:20 2 -73.962982177734375 40.756679534912109 -73.984405517578125 40.760719299316406 N 1114 id0799785 2 2016-06-01 20:58:29 2016-06-01 21:02:49 1 -73.956306457519531 40.767940521240234 -73.966110229492188 40.76300048828125 N 260 id2900608 2 2016-05-27 00:43:36 2016-05-27 01:07:10 1 -73.992195129394531 40.727226257324219 -73.974655151367188 40.783069610595703 N 1414 id3319787 1 2016-05-16 15:29:02 2016-05-16 15:32:33 1 -73.955513000488281 40.768592834472656 -73.948760986328125 40.77154541015625 N 211 id3379579 2 2016-04-11 17:29:50 2016-04-11 18:08:26 1 -73.991165161132813 40.755561828613281 -73.999290466308594 40.725353240966797 N 2316 id1154431 1 2016-04-14 08:48:26 2016-04-14 09:00:37 1 -73.994255065917969 40.745803833007813 -73.999656677246094 40.723342895507813 N 731 id3552682 1 2016-06-27 09:55:13 2016-06-27 10:17:10 1 -74.003982543945313 40.7130126953125 -73.979194641113281 40.749923706054688 N 1317 id3390316 2 2016-06-05 13:47:23 2016-06-05 13:51:34 1 -73.98388671875 40.738197326660156 -73.991203308105469 40.727870941162109 N 251 id2070428 1 2016-02-28 02:23:02 2016-02-28 02:31:08 1 -73.980369567871094 40.742420196533203 -73.962852478027344 40.760635375976563 N 486 id0809232 2 2016-04-01 12:12:25 2016-04-01 12:23:17 1 -73.979537963867188 40.753360748291016 -73.963996887207031 40.763458251953125 N 652 id2352683 1 2016-04-09 03:34:27 2016-04-09 03:41:30 1 -73.995864868164063 40.758811950683594 -73.993324279785156 40.740322113037109 N 423 id1603037 1 2016-06-25 10:36:26 2016-06-25 10:55:49 1 -73.993553161621094 40.747173309326172 -74.006141662597656 40.704383850097656 N 1163 id3321406 2 2016-06-03 08:15:05 2016-06-03 08:56:30 1 -73.955230712890625 40.777133941650391 -73.788749694824219 40.641471862792969 N 2485 id0129640 2 2016-02-14 13:27:56 2016-02-14 13:49:19 1 -73.956581115722656 40.771358489990234 -73.974967956542969 40.732791900634766 N 1283 id3587298 1 2016-02-27 21:56:01 2016-02-27 22:14:51 1 -73.9837646484375 40.749874114990234 -73.958831787109375 40.800960540771484 N 1130 id2104175 1 2016-06-20 23:07:16 2016-06-20 23:18:50 1 -73.95843505859375 40.713191986083984 -73.949539184570312 40.6802

In [5]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.

#permanent_table_name = "yellow_tripdata_2017_01_8bceb_csv"

# df.write.format("parquet").saveAsTable(permanent_table_name)

In [6]:
#df.select('id').distinct().show()

Count of number of unique entry

In [8]:
from pyspark.sql.functions import col, countDistinct

df.agg(countDistinct(col("id")).alias("count")).show()

+-------+
 count|
+-------+
1458644|
+-------+

Count number of null or nan value in each column

In [10]:
from pyspark.sql.functions import isnan, when, count, col
df.select([count(when(isnan(c), c)).alias(c) for c in df.columns]).show()

+---+---------+---------------+----------------+---------------+----------------+---------------+-----------------+----------------+------------------+-------------+
 id|vendor_id|pickup_datetime|dropoff_datetime|passenger_count|pickup_longitude|pickup_latitude|dropoff_longitude|dropoff_latitude|store_and_fwd_flag|trip_duration|
+---+---------+---------------+----------------+---------------+----------------+---------------+-----------------+----------------+------------------+-------------+
 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0|
+---+---------+---------------+----------------+---------------+----------------+---------------+-----------------+----------------+------------------+-------------+

In [11]:
df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: string (nullable = true)
-- dropoff_datetime: string (nullable = true)
-- passenger_count: string (nullable = true)
-- pickup_longitude: string (nullable = true)
-- pickup_latitude: string (nullable = true)
-- dropoff_longitude: string (nullable = true)
-- dropoff_latitude: string (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: string (nullable = true)

In [12]:
df = df.withColumn("pickup_datetime", df["pickup_datetime"].cast("timestamp"))
df = df.withColumn("dropoff_datetime", df["dropoff_datetime"].cast("timestamp"))
df = df.withColumn("passenger_count", df["passenger_count"].cast("int"))
df = df.withColumn("trip_duration", df["trip_duration"].cast("float"))
df = df.withColumn("pickup_longitude", df["pickup_longitude"].cast("float"))
df = df.withColumn("pickup_latitude", df["pickup_latitude"].cast("float"))
df = df.withColumn("dropoff_longitude", df["dropoff_longitude"].cast("float"))
df = df.withColumn("dropoff_latitude", df["dropoff_latitude"].cast("float"))

In [13]:
df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: timestamp (nullable = true)
-- dropoff_datetime: timestamp (nullable = true)
-- passenger_count: integer (nullable = true)
-- pickup_longitude: float (nullable = true)
-- pickup_latitude: float (nullable = true)
-- dropoff_longitude: float (nullable = true)
-- dropoff_latitude: float (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: float (nullable = true)

In [14]:
display(df)

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration id2875421 2 2016-03-14T17:24:55.000+0000 2016-03-14T17:32:30.000+0000 1 -73.982155 40.767937 -73.96463 40.765602 N 455.0 id2377394 1 2016-06-12T00:43:35.000+0000 2016-06-12T00:54:38.000+0000 1 -73.980415 40.738564 -73.99948 40.73115 N 663.0 id3858529 2 2016-01-19T11:35:24.000+0000 2016-01-19T12:10:48.000+0000 1 -73.97903 40.76394 -74.00533 40.710087 N 2124.0 id3504673 2 2016-04-06T19:32:31.000+0000 2016-04-06T19:39:40.000+0000 1 -74.01004 40.71997 -74.01227 40.70672 N 429.0 id2181028 2 2016-03-26T13:30:55.000+0000 2016-03-26T13:38:10.000+0000 1 -73.97305 40.79321 -73.97292 40.78252 N 435.0 id0801584 2 2016-01-30T22:01:40.000+0000 2016-01-30T22:09:03.000+0000 6 -73.98286 40.742195 -73.99208 40.749184 N 443.0 id1813257 1 2016-06-17T22:34:59.000+0000 2016-06-17T22:40:40.000+0000 4 -73.96902 40.75784 -73.957405 40.765896 N 341.0 id1324603 2 2016-05-21T07:54:58.000+0000 2016-05-21T08:20:49.000+0000 1 -73.96928 40.79778 -73.92247 40.76056 N 1551.0 id1301050 1 2016-05-27T23:12:23.000+0000 2016-05-27T23:16:38.000+0000 1 -73.99948 40.7384 -73.98579 40.732815 N 255.0 id0012891 2 2016-03-10T21:45:01.000+0000 2016-03-10T22:05:26.000+0000 1 -73.98105 40.74434 -73.973 40.78999 N 1225.0 id1436371 2 2016-05-10T22:08:41.000+0000 2016-05-10T22:29:55.000+0000 1 -73.98265 40.76384 -74.00223 40.73299 N 1274.0 id1299289 2 2016-05-15T11:16:11.000+0000 2016-05-15T11:34:59.000+0000 4 -73.99153 40.74944 -73.95654 40.77063 N 1128.0 id1187965 2 2016-02-19T09:52:46.000+0000 2016-02-19T10:11:20.000+0000 2 -73.96298 40.75668 -73.984406 40.76072 N 1114.0 id0799785 2 2016-06-01T20:58:29.000+0000 2016-06-01T21:02:49.000+0000 1 -73.95631 40.76794 -73.96611 40.763 N 260.0 id2900608 2 2016-05-27T00:43:36.000+0000 2016-05-27T01:07:10.000+0000 1 -73.992195 40.727226 -73.974655 40.78307 N 1414.0 id3319787 1 2016-05-16T15:29:02.000+0000 2016-05-16T15:32:33.000+0000 1 -73.95551 40.768593 -73.94876 40.771545 N 211.0 id3379579 2 2016-04-11T17:29:50.000+0000 2016-04-11T18:08:26.000+0000 1 -73.991165 40.75556 -73.99929 40.725353 N 2316.0 id1154431 1 2016-04-14T08:48:26.000+0000 2016-04-14T09:00:37.000+0000 1 -73.994255 40.745804 -73.99966 40.723343 N 731.0 id3552682 1 2016-06-27T09:55:13.000+0000 2016-06-27T10:17:10.000+0000 1 -74.00398 40.713013 -73.979195 40.749924 N 1317.0 id3390316 2 2016-06-05T13:47:23.000+0000 2016-06-05T13:51:34.000+0000 1 -73.98389 40.738197 -73.9912 40.72787 N 251.0 id2070428 1 2016-02-28T02:23:02.000+0000 2016-02-28T02:31:08.000+0000 1 -73.98037 40.74242 -73.96285 40.760635 N 486.0 id0809232 2 2016-04-01T12:12:25.000+0000 2016-04-01T12:23:17.000+0000 1 -73.97954 40.75336 -73.964 40.76346 N 652.0 id2352683 1 2016-04-09T03:34:27.000+0000 2016-04-09T03:41:30.000+0000 1 -73.995865 40.758812 -73.993324 40.740322 N 423.0 id1603037 1 2016-06-25T10:36:26.000+0000 2016-06-25T10:55:49.000+0000 1 -73.99355 40.747173 -74.00614 40.704384 N 1163.0 id3321406 2 2016-06-03T08:15:05.000+0000 2016-06-03T08:56:30.000+0000 1 -73.95523 40.777134 -73.78875 40.64147 N 2485.0 id0129640 2 2016-02-14T13:27:56.000+0000 2016-02-14T13:49:19.000+0000 1 -73.95658 40.77136 -73.97497 40.73279 N 1283.0 id3587298 1 2016-02-27T21:56:01.000+0000 2016-02-27T22:14:51.000+0000 1 -73.983765 40.749874 -73.95883 40.80096 N 1130.0 id2104175 1 2016-06-20T23:07:16.000+0000 2016-06-20T23:18:50.000+0000 1 -73.958435 40.713192 -73.94954 40.680252 N 694.0 id3973319 2 2016-06-13T21:57:27.000+0000 2016-06-13T22:12:19.000+0000 1 -73.99422 40.713306 -73.98285 40.6923 N 892.0 id1410897 1 2016-03-23T14:10:39.000+0000 2016-03-23T14:49:30.000+0000 1 -73.98212 40.75635 -73.86569 40.77099 N 2331.0 id0266980 1 2016-05-17T10:47:51.000+0000 2016-05-17T11:12:30.000+0000 1 -73.97094 40.76167 -73.99624 40.72537 N 1479.0 id2822549 2 2016-05-21T13:03:24.000+0000 2016-05-21T13:20:52.000+0000 1 -73.99827 40.72004 -74.01036 40.67214 N 1048.0 id3075902 2 20

In [15]:
from pyspark.sql.functions import *
df = df.withColumn("weekday", date_format(df.pickup_datetime, 'EEEE'))
df = df.withColumn("weekday_num", dayofweek(df.pickup_datetime))
df = df.withColumn("month", month(df.pickup_datetime))
df = df.withColumn("pickup_hour", hour(df.pickup_datetime))

In [16]:
df.select('weekday').distinct().show()

+---------+
 weekday|
+---------+
Wednesday|
 Tuesday|
 Friday|
 Thursday|
 Saturday|
 Monday|
 Sunday|
+---------+

Haversine formula

In [18]:
df = df.withColumn("a", pow(sin(toRadians(df.dropoff_latitude - df.pickup_latitude) / 2), 2) + cos(toRadians(df.pickup_latitude)) * cos(toRadians(df.dropoff_latitude)) * pow(sin(toRadians(df.dropoff_longitude - df.pickup_longitude) / 2), 2))#.withColumn("distance", atan2(sqrt(df.a), sqrt(-df.a + 1)) * 2 * 6371)

/databricks/spark/python/pyspark/sql/functions.py:54: DeprecationWarning: Deprecated in 2.1, use radians instead.
 warnings.warn(message, DeprecationWarning)

In [19]:
df = df.withColumn("distance", round(atan2(sqrt(df.a), sqrt(-df.a + 1)) * 2 * 6371,2))

In [20]:
#Calculate Speed in km/h for further insights
df = df.withColumn("speed", round(df.distance/(df.trip_duration/3600),2))

In [21]:
display(df)

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration weekday weekday_num month pickup_hour a distance speed id2875421 2 2016-03-14T17:24:55.000+0000 2016-03-14T17:32:30.000+0000 1 -73.982155 40.767937 -73.96463 40.765602 N 455.0 Monday 2 3 17 1.3830896636179652E-8 1.5 11.87 id2377394 1 2016-06-12T00:43:35.000+0000 2016-06-12T00:54:38.000+0000 1 -73.980415 40.738564 -73.99948 40.73115 N 663.0 Sunday 1 6 0 2.0078128522717423E-8 1.81 9.83 id3858529 2 2016-01-19T11:35:24.000+0000 2016-01-19T12:10:48.000+0000 1 -73.97903 40.76394 -74.00533 40.710087 N 2124.0 Tuesday 3 1 11 2.511076618143245E-7 6.39 10.83 id3504673 2 2016-04-06T19:32:31.000+0000 2016-04-06T19:39:40.000+0000 1 -74.01004 40.71997 -74.01227 40.70672 N 429.0 Wednesday 4 4 19 1.359155624250055E-8 1.49 12.5 id2181028 2 2016-03-26T13:30:55.000+0000 2016-03-26T13:38:10.000+0000 1 -73.97305 40.79321 -73.97292 40.78252 N 435.0 Saturday 7 3 13 8.701373594095704E-9 1.19 9.85 id0801584 2 2016-01-30T22:01:40.000+0000 2016-01-30T22:09:03.000+0000 6 -73.98286 40.742195 -73.99208 40.749184 N 443.0 Saturday 7 1 22 7.438317364467243E-9 1.1 8.94 id1813257 1 2016-06-17T22:34:59.000+0000 2016-06-17T22:40:40.000+0000 4 -73.96902 40.75784 -73.957405 40.765896 N 341.0 Friday 6 6 22 1.0834136581882085E-8 1.33 14.04 id1324603 2 2016-05-21T07:54:58.000+0000 2016-05-21T08:20:49.000+0000 1 -73.96928 40.79778 -73.92247 40.76056 N 1551.0 Saturday 7 5 7 2.0116586779003735E-7 5.71 13.25 id1301050 1 2016-05-27T23:12:23.000+0000 2016-05-27T23:16:38.000+0000 1 -73.99948 40.7384 -73.98579 40.732815 N 255.0 Friday 6 5 23 1.0575516792025634E-8 1.31 18.49 id0012891 2 2016-03-10T21:45:01.000+0000 2016-03-10T22:05:26.000+0000 1 -73.98105 40.74434 -73.973 40.78999 N 1225.0 Thursday 5 3 21 1.6153317501260625E-7 5.12 15.05 id1436371 2 2016-05-10T22:08:41.000+0000 2016-05-10T22:29:55.000+0000 1 -73.98265 40.76384 -74.00223 40.73299 N 1274.0 Tuesday 3 5 22 8.922656314182737E-8 3.81 10.77 id1299289 2 2016-05-15T11:16:11.000+0000 2016-05-15T11:34:59.000+0000 4 -73.99153 40.74944 -73.95654 40.77063 N 1128.0 Sunday 1 5 11 8.768402598600243E-8 3.77 12.03 id1187965 2 2016-02-19T09:52:46.000+0000 2016-02-19T10:11:20.000+0000 2 -73.96298 40.75668 -73.984406 40.76072 N 1114.0 Friday 6 2 9 2.129654877938154E-8 1.86 6.01 id0799785 2 2016-06-01T20:58:29.000+0000 2016-06-01T21:02:49.000+0000 1 -73.95631 40.76794 -73.96611 40.763 N 260.0 Wednesday 4 6 20 6.057203368798397E-9 0.99 13.71 id2900608 2 2016-05-27T00:43:36.000+0000 2016-05-27T01:07:10.000+0000 1 -73.992195 40.727226 -73.974655 40.78307 N 1414.0 Friday 6 5 0 2.509297192595758E-7 6.38 16.24 id3319787 1 2016-05-16T15:29:02.000+0000 2016-05-16T15:32:33.000+0000 1 -73.95551 40.768593 -73.94876 40.771545 N 211.0 Monday 2 5 15 2.655201508351016E-9 0.66 11.26 id3379579 2 2016-04-11T17:29:50.000+0000 2016-04-11T18:08:26.000+0000 1 -73.991165 40.75556 -73.99929 40.725353 N 2316.0 Monday 2 4 17 7.238160272905697E-8 3.43 5.33 id1154431 1 2016-04-14T08:48:26.000+0000 2016-04-14T09:00:37.000+0000 1 -73.994255 40.745804 -73.99966 40.723343 N 731.0 Thursday 5 4 8 3.9695192884170783E-8 2.54 12.51 id3552682 1 2016-06-27T09:55:13.000+0000 2016-06-27T10:17:10.000+0000 1 -74.00398 40.713013 -73.979195 40.749924 N 1317.0 Monday 2 6 9 1.3062369794697426E-7 4.61 12.6 id3390316 2 2016-06-05T13:47:23.000+0000 2016-06-05T13:51:34.000+0000 1 -73.98389 40.738197 -73.9912 40.72787 N 251.0 Sunday 1 6 13 1.0461510732205175E-8 1.3 18.65 id2070428 1 2016-02-28T02:23:02.000+0000 2016-02-28T02:31:08.000+0000 1 -73.98037 40.74242 -73.96285 40.760635 N 486.0 Sunday 1 2 2 3.867776267452501E-8 2.51 18.59 id0809232 2 2016-04-01T12:12:25.000+0000 2016-04-01T12:23:17.000+0000 1 -73.97954 40.75336 -73.964 40.76346 N 652.0 Friday 6 4 12 1.8317931467470727E-8 1.72 9.5 id2352683 1 2016-04-09T03:34:27.000+0000 2016-04-09T03:41:30.000+0000 1 -73.995865 40.758812 -73.993324 40.740322 N 423.0 Saturday 7 4 3 

In [22]:
df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: timestamp (nullable = true)
-- dropoff_datetime: timestamp (nullable = true)
-- passenger_count: integer (nullable = true)
-- pickup_longitude: float (nullable = true)
-- pickup_latitude: float (nullable = true)
-- dropoff_longitude: float (nullable = true)
-- dropoff_latitude: float (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: float (nullable = true)
-- weekday: string (nullable = true)
-- weekday_num: integer (nullable = true)
-- month: integer (nullable = true)
-- pickup_hour: integer (nullable = true)
-- a: double (nullable = true)
-- distance: double (nullable = true)
-- speed: double (nullable = true)

In [23]:
cols = ['passenger_count','month','weekday_num','pickup_hour']
for col_name in cols:
  df = df.withColumn(col_name, df[col_name].cast('string'))

df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: timestamp (nullable = true)
-- dropoff_datetime: timestamp (nullable = true)
-- passenger_count: string (nullable = true)
-- pickup_longitude: float (nullable = true)
-- pickup_latitude: float (nullable = true)
-- dropoff_longitude: float (nullable = true)
-- dropoff_latitude: float (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: float (nullable = true)
-- weekday: string (nullable = true)
-- weekday_num: string (nullable = true)
-- month: string (nullable = true)
-- pickup_hour: string (nullable = true)
-- a: double (nullable = true)
-- distance: double (nullable = true)
-- speed: double (nullable = true)

In [24]:
df.createOrReplaceTempView(temp_table_name)

In [25]:
from pyspark.sql.functions import count
df.groupBy('passenger_count').count().show()

+---------------+-------+
passenger_count| count|
+---------------+-------+
 7| 3|
 3| 59896|
 8| 1|
 0| 60|
 5| 78088|
 6| 48333|
 9| 1|
 1|1033540|
 4| 28404|
 2| 210318|
+---------------+-------+

In [26]:
%sql
select distinct passenger_count, count(passenger_count) as Count from `train_csv` group by passenger_count order by passenger_count

passenger_count Count 0 60 1 1033540 2 210318 3 59896 4 28404 5 78088 6 48333 7 3 8 1 9 1

In [27]:
df.describe('passenger_count').show()

+-------+------------------+
summary| passenger_count|
+-------+------------------+
 count| 1458644|
 mean|1.6645295219395548|
 stddev| 1.314242167823115|
 min| 0|
 max| 9|
+-------+------------------+

Replacing 0 by 2 passengers as mean is 1.66

In [29]:
from pyspark.sql import functions as F
df = df.withColumn("passenger_count", F.when(F.col("passenger_count")== 0, 2).otherwise(F.col("passenger_count")))

In [30]:
from pyspark.sql.functions import count
df.groupBy('passenger_count').count().show()

+---------------+-------+
passenger_count| count|
+---------------+-------+
 7| 3|
 3| 59896|
 8| 1|
 5| 78088|
 6| 48333|
 9| 1|
 1|1033540|
 4| 28404|
 2| 210378|
+---------------+-------+

Delete passanger 7,8 and 9

In [32]:
df = df.filter(col("passenger_count") != "7")

In [33]:
df = df.filter(col("passenger_count") != "8").filter(col("passenger_count") != "9")

In [34]:
df.groupBy('passenger_count').count().show()

+---------------+-------+
passenger_count| count|
+---------------+-------+
 3| 59896|
 5| 78088|
 6| 48333|
 1|1033540|
 4| 28404|
 2| 210378|
+---------------+-------+

In [35]:
df.createOrReplaceTempView(temp_table_name)

In [36]:
%sql
select distinct passenger_count, count(passenger_count) as Count from `train_csv` group by passenger_count order by passenger_count

passenger_count Count 1 1033540 2 210378 3 59896 4 28404 5 78088 6 48333

In [37]:
%sql
select distinct vendor_id, count(vendor_id) as Count from `train_csv` group by vendor_id order by vendor_id

vendor_id Count 1 678342 2 780297

In [38]:
df.describe('distance').show()

+-------+------------------+
summary| distance|
+-------+------------------+
 count| 1458639|
 mean|3.4408716481597104|
 stddev| 4.296545964294174|
 min| 0.0|
 max| 1240.91|
+-------+------------------+

In [39]:
%sql
select distinct distance, count(distance) as Count from `train_csv` group by distance order by distance

distance Count 0.0 7931 0.01 1256 0.02 729 0.03 632 0.04 530 0.05 495 0.06 385 0.07 413 0.08 421 0.09 353 0.1 328 0.11 313 0.12 337 0.13 330 0.14 341 0.15 354 0.16 393 0.17 396 0.18 341 0.19 379 0.2 409 0.21 431 0.22 472 0.23 509 0.24 548 0.25 561 0.26 563 0.27 595 0.28 667 0.29 682 0.3 760 0.31 778 0.32 827 0.33 937 0.34 897 0.35 994 0.36 1072 0.37 1032 0.38 1060 0.39 1143 0.4 1307 0.41 1359 0.42 1465 0.43 1531 0.44 1600 0.45 1631 0.46 1738 0.47 1804 0.48 1843 0.49 1973 0.5 2075 0.51 2096 0.52 2236 0.53 2322 0.54 2491 0.55 2523 0.56 2764 0.57 2663 0.58 2773 0.59 2946 0.6 2935 0.61 3016 0.62 3174 0.63 3211 0.64 3259 0.65 3339 0.66 3470 0.67 3589 0.68 3589 0.69 3563 0.7 3923 0.71 3800 0.72 3998 0.73 4056 0.74 3993 0.75 4233 0.76 4235 0.77 4225 0.78 4263 0.79 4437 0.8 4417 0.81 4510 0.82 4573 0.83 4645 0.84 4844 0.85 4761 0.86 4731 0.87 4848 0.88 4935 0.89 4895 0.9 4877 0.91 4799 0.92 4935 0.93 4933 0.94 4877 0.95 4888 0.96 4977 0.97 4919 0.98 5116 0.99 4978 1.0 4890 1.01 5007 1.02 5200 1.03 5110 1.04 5131 1.05 5205 1.06 5110 1.07 5114 1.08 5136 1.09 5081 1.1 4852 1.11 5178 1.12 5068 1.13 5239 1.14 5114 1.15 5157 1.16 5166 1.17 5125 1.18 5254 1.19 5067 1.2 5255 1.21 5004 1.22 5195 1.23 5097 1.24 5032 1.25 4971 1.26 4900 1.27 4944 1.28 4918 1.29 4911 1.3 4931 1.31 4827 1.32 5079 1.33 4808 1.34 4968 1.35 5066 1.36 4775 1.37 4814 1.38 4954 1.39 4968 1.4 4746 1.41 4675 1.42 4586 1.43 4632 1.44 4674 1.45 4678 1.46 4631 1.47 4559 1.48 4609 1.49 4608 1.5 4584 1.51 4564 1.52 4460 1.53 4467 1.54 4407 1.55 4473 1.56 4449 1.57 4409 1.58 4293 1.59 4360 1.6 4292 1.61 4363 1.62 4301 1.63 4255 1.64 4285 1.65 4254 1.66 4348 1.67 4070 1.68 4225 1.69 4270 1.7 4246 1.71 4171 1.72 4183 1.73 4137 1.74 4081 1.75 4098 1.76 4097 1.77 4069 1.78 4117 1.79 4010 1.8 3828 1.81 4052 1.82 3945 1.83 3868 1.84 3890 1.85 3893 1.86 3848 1.87 3839 1.88 3832 1.89 3688 1.9 3677 1.91 3726 1.92 3721 1.93 3720 1.94 3695 1.95 3603 1.96 3744 1.97 3716 1.98 3661 1.99 3556 2.0 3519 2.01 3669 2.02 3622 2.03 3549 2.04 3404 2.05 3325 2.06 3376 2.07 3352 2.08 3330 2.09 3239 2.1 3268 2.11 3327 2.12 3381 2.13 3336 2.14 3248 2.15 3334 2.16 3218 2.17 3287 2.18 3226 2.19 3193 2.2 3127 2.21 3132 2.22 3071 2.23 3039 2.24 3034 2.25 3132 2.26 3052 2.27 3007 2.28 2961 2.29 2928 2.3 2930 2.31 2992 2.32 2838 2.33 2919 2.34 2924 2.35 2919 2.36 2835 2.37 2869 2.38 2867 2.39 2768 2.4 2830 2.41 2756 2.42 2783 2.43 2701 2.44 2783 2.45 2728 2.46 2609 2.47 2675 2.48 2659 2.49 2606 2.5 2603 2.51 2570 2.52 2622 2.53 2515 2.54 2598 2.55 2524 2.56 2591 2.57 2524 2.58 2443 2.59 2541 2.6 2522 2.61 2518 2.62 2496 2.63 2352 2.64 2441 2.65 2354 2.66 2331 2.67 2356 2.68 2240 2.69 2324 2.7 2308 2.71 2344 2.72 2280 2.73 2296 2.74 2208 2.75 2199 2.76 2242 2.77 2180 2.78 2144 2.79 2208 2.8 2192 2.81 2085 2.82 2052 2.83 2121 2.84 2062 2.85 2139 2.86 2112 2.87 2169 2.88 2046 2.89 1977 2.9 1975 2.91 2087 2.92 2015 2.93 1991 2.94 2025 2.95 1971 2.96 1966 2.97 1925 2.98 1950 2.99 2018 3.0 1936 3.01 1923 3.02 1945 3.03 1859 3.04 1817 3.05 1869 3.06 1895 3.07 1775 3.08 1792 3.09 1851 3.1 1804 3.11 1798 3.12 1755 3.13 1797 3.14 1724 3.15 1733 3.16 1734 3.17 1705 3.18 1676 3.19 1722 3.2 1661 3.21 1730 3.22 1718 3.23 1633 3.24 1643 3.25 1678 3.26 1635 3.27 1638 3.28 1614 3.29 1587 3.3 1681 3.31 1555 3.32 1530 3.33 1576 3.34 1609 3.35 1513 3.36 1584 3.37 1543 3.38 1441 3.39 1549 3.4 1433 3.41 1533 3.42 1490 3.43 1350 3.44 1440 3.45 1513 3.46 1496 3.47 1430 3.48 1432 3.49 1427 3.5 1446 3.51 1439 3.52 1424 3.53 1486 3.54 1373 3.55 1423 3.56 1346 3.57 1369 3.58 1428 3.59 1422 3.6 1359 3.61 1301 3.62 1305 3.63 1401 3.64 1327 3.65 1271 3.66 1353 3.67 1369 3.68 1303 3.69 1335 3.7 1267 3.71 1306 3.72 1294 3.73 1238 3.74 1200 3.75 1248 3.76 1230 3.77 1237 3.78 1291 3.79 1208 3.8 1119 3.81 1146 3.82 1290 3.83 1216 3.84 1158 3.85 1126 3.86 1137 3.87 1185 3.88 1169 3.89 1179 3.9 1124 3.91 1096 3.92 1143 3.93 1131 3.94 1133 3.95 1122 3.96 1108 3.97 1151 3.98 1106 3.99 1061 4.0 1032 4.01 1085 4.02 1097 4.03 1052 4.04 1019 4.05 1096 

In [40]:
df.describe('trip_duration').show()

+-------+-----------------+
summary| trip_duration|
+-------+-----------------+
 count| 1458639|
 mean|959.4950662912482|
 stddev|5237.440469291621|
 min| 1.0|
 max| 3526282.0|
+-------+-----------------+

In [41]:
df = df.withColumn("trip_duration", round(df.trip_duration/60,2))

In [42]:
df.describe('trip_duration').show()

+-------+------------------+
summary| trip_duration|
+-------+------------------+
 count| 1458639|
 mean|15.991585608228014|
 stddev| 87.29067584757281|
 min| 0.02|
 max| 58771.37|
+-------+------------------+

In [43]:
df.createOrReplaceTempView(temp_table_name)

In [44]:
%sql
select distinct trip_duration, count(trip_duration) as Count from `train_csv` group by trip_duration order by trip_duration

trip_duration Count 0.02 33 0.03 177 0.05 318 0.07 317 0.08 284 0.1 192 0.12 218 0.13 206 0.15 239 0.17 182 0.18 164 0.2 152 0.22 162 0.23 176 0.25 147 0.27 126 0.28 149 0.3 128 0.32 149 0.33 127 0.35 122 0.37 125 0.38 122 0.4 140 0.42 139 0.43 116 0.45 123 0.47 122 0.48 124 0.5 108 0.52 151 0.53 100 0.55 130 0.57 122 0.58 144 0.6 94 0.62 111 0.63 104 0.65 100 0.67 132 0.68 108 0.7 116 0.72 121 0.73 122 0.75 132 0.77 101 0.78 141 0.8 126 0.82 141 0.83 147 0.85 118 0.87 121 0.88 122 0.9 145 0.92 136 0.93 149 0.95 168 0.97 148 0.98 155 1.0 182 1.02 171 1.03 176 1.05 189 1.07 181 1.08 191 1.1 170 1.12 152 1.13 174 1.15 194 1.17 215 1.18 227 1.2 189 1.22 223 1.23 242 1.25 255 1.27 236 1.28 210 1.3 219 1.32 254 1.33 258 1.35 241 1.37 242 1.38 279 1.4 258 1.42 311 1.43 297 1.45 261 1.47 336 1.48 322 1.5 349 1.52 332 1.53 332 1.55 320 1.57 340 1.58 391 1.6 396 1.62 291 1.63 351 1.65 362 1.67 375 1.68 393 1.7 402 1.72 379 1.73 367 1.75 448 1.77 476 1.78 393 1.8 414 1.82 459 1.83 458 1.85 439 1.87 459 1.88 445 1.9 529 1.92 453 1.93 554 1.95 476 1.97 485 1.98 498 2.0 501 2.02 549 2.03 523 2.05 535 2.07 567 2.08 557 2.1 612 2.12 498 2.13 537 2.15 569 2.17 573 2.18 635 2.2 641 2.22 542 2.23 611 2.25 671 2.27 725 2.28 564 2.3 627 2.32 633 2.33 636 2.35 649 2.37 691 2.38 656 2.4 664 2.42 745 2.43 805 2.45 642 2.47 684 2.48 735 2.5 779 2.52 733 2.53 805 2.55 724 2.57 715 2.58 815 2.6 853 2.62 760 2.63 767 2.65 797 2.67 839 2.68 813 2.7 911 2.72 808 2.73 811 2.75 840 2.77 956 2.78 910 2.8 825 2.82 895 2.83 933 2.85 935 2.87 928 2.88 864 2.9 904 2.92 929 2.93 1000 2.95 955 2.97 847 2.98 959 3.0 977 3.02 1001 3.03 996 3.05 996 3.07 979 3.08 1016 3.1 1024 3.12 1034 3.13 883 3.15 965 3.17 980 3.18 975 3.2 1020 3.22 1034 3.23 1005 3.25 1077 3.27 1066 3.28 1117 3.3 956 3.32 1072 3.33 1117 3.35 1105 3.37 1123 3.38 1058 3.4 1024 3.42 1044 3.43 1161 3.45 1121 3.47 1060 3.48 1004 3.5 1087 3.52 1140 3.53 1181 3.55 1157 3.57 1106 3.58 1171 3.6 1205 3.62 1263 3.63 1073 3.65 1145 3.67 1262 3.68 1176 3.7 1191 3.72 1227 3.73 1131 3.75 1264 3.77 1170 3.78 1320 3.8 1150 3.82 1203 3.83 1211 3.85 1186 3.87 1222 3.88 1309 3.9 1164 3.92 1169 3.93 1301 3.95 1319 3.97 1174 3.98 1140 4.0 1212 4.02 1275 4.03 1291 4.05 1209 4.07 1231 4.08 1277 4.1 1272 4.12 1359 4.13 1273 4.15 1194 4.17 1291 4.18 1313 4.2 1381 4.22 1348 4.23 1293 4.25 1308 4.27 1401 4.28 1442 4.3 1375 4.32 1300 4.33 1332 4.35 1377 4.37 1404 4.38 1290 4.4 1370 4.42 1297 4.43 1445 4.45 1395 4.47 1281 4.48 1311 4.5 1331 4.52 1305 4.53 1400 4.55 1404 4.57 1327 4.58 1351 4.6 1410 4.62 1488 4.63 1368 4.65 1300 4.67 1343 4.68 1456 4.7 1393 4.72 1440 4.73 1381 4.75 1408 4.77 1430 4.78 1519 4.8 1430 4.82 1348 4.83 1360 4.85 1460 4.87 1396 4.88 1460 4.9 1423 4.92 1417 4.93 1386 4.95 1473 4.97 1486 4.98 1434 5.0 1431 5.02 1402 5.03 1446 5.05 1454 5.07 1420 5.08 1378 5.1 1427 5.12 1485 5.13 1465 5.15 1426 5.17 1340 5.18 1366 5.2 1409 5.22 1488 5.23 1441 5.25 1467 5.27 1426 5.28 1528 5.3 1506 5.32 1412 5.33 1342 5.35 1446 5.37 1469 5.38 1503 5.4 1511 5.42 1519 5.43 1462 5.45 1560 5.47 1493 5.48 1394 5.5 1341 5.52 1480 5.53 1538 5.55 1469 5.57 1556 5.58 1510 5.6 1451 5.62 1530 5.63 1494 5.65 1465 5.67 1310 5.68 1456 5.7 1423 5.72 1506 5.73 1561 5.75 1471 5.77 1493 5.78 1426 5.8 1582 5.82 1455 5.83 1472 5.85 1439 5.87 1549 5.88 1539 5.9 1538 5.92 1482 5.93 1437 5.95 1542 5.97 1577 5.98 1493 6.0 1510 6.02 1453 6.03 1540 6.05 1521 6.07 1526 6.08 1457 6.1 1487 6.12 1581 6.13 1624 6.15 1556 6.17 1532 6.18 1430 6.2 1515 6.22 1422 6.23 1514 6.25 1533 6.27 1539 6.28 1531 6.3 1550 6.32 1486 6.33 1457 6.35 1391 6.37 1509 6.38 1450 6.4 1529 6.42 1457 6.43 1538 6.45 1448 6.47 1567 6.48 1476 6.5 1501 6.52 1399 6.53 1481 6.55 1509 6.57 1491 6.58 1525 6.6 1418 6.62 1425 6.63 1536 6.65 1573 6.67 1476 6.68 1420 6.7 1444 6.72 1540 6.73 1533 6.75 1511 6.77 1518 6.78 1413 6.8 1584 6.82 1524 6.83 1481 6.85 1394 6.87 1470 6.88 1506 6.9 1464 6.92 1484 6.93 1532 6.95 1568 6.97 1570 6.98 1523 7.0 1530 7.02 1484 7.03 1550 7.05 1504 7.07 1547

In [45]:
%sql
select * from `train_csv` where trip_duration >= 900

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration weekday weekday_num month pickup_hour a distance speed id3307903 2 2016-02-20T04:03:06.000+0000 2016-02-21T03:33:00.000+0000 3 -74.0081 40.74149 -74.00996 40.71461 N 1409.9 Saturday 7 2 4 5.5167732515428266E-8 2.99 0.13 id1091477 2 2016-05-07T18:36:22.000+0000 2016-05-08T18:32:11.000+0000 1 -73.99024 40.75092 -73.97628 40.75089 N 1435.82 Saturday 7 5 18 8.519393142812997E-9 1.18 0.05 id3431345 2 2016-06-07T12:58:48.000+0000 2016-06-08T12:58:00.000+0000 6 -73.954956 40.77765 -73.98103 40.743713 N 1439.2 Tuesday 3 6 12 1.174121079743924E-7 4.37 0.18 id1487069 2 2016-02-13T00:21:49.000+0000 2016-02-14T00:19:05.000+0000 1 -73.96859 40.799217 -73.97958 40.784714 N 1437.27 Saturday 7 2 0 2.129500678117729E-8 1.86 0.08 id3674870 2 2016-03-18T11:54:20.000+0000 2016-03-19T11:34:17.000+0000 1 -73.98909 40.73699 -73.972336 40.75151 N 1419.95 Friday 6 3 11 2.832314020525033E-8 2.14 0.09 id3632390 2 2016-06-08T08:54:33.000+0000 2016-06-09T07:58:09.000+0000 5 -73.98056 40.742466 -73.98472 40.74861 N 1383.6 Wednesday 4 6 8 3.631841993406624E-9 0.77 0.03 id3354426 2 2016-05-05T15:18:41.000+0000 2016-05-06T15:11:07.000+0000 1 -73.98904 40.773514 -73.98068 40.78152 N 1432.43 Thursday 5 5 15 7.93024990478118E-9 1.13 0.05 id0773526 2 2016-04-02T14:58:45.000+0000 2016-04-03T14:19:55.000+0000 6 -73.98799 40.76134 -74.00292 40.75628 N 1401.17 Saturday 7 4 14 1.1691946367015729E-8 1.38 0.06 id3617210 2 2016-03-15T17:51:32.000+0000 2016-03-16T17:18:04.000+0000 1 -73.96561 40.76578 -73.97265 40.753418 N 1406.53 Tuesday 3 3 17 1.380720493346012E-8 1.5 0.06 id0067152 2 2016-02-27T21:04:05.000+0000 2016-02-28T21:03:22.000+0000 5 -73.993744 40.727444 -74.001335 40.729244 N 1439.28 Saturday 7 2 21 2.767137309897393E-9 0.67 0.03 id3013398 2 2016-05-17T12:43:48.000+0000 2016-05-18T12:38:03.000+0000 2 -74.002625 40.73362 -73.994484 40.750603 N 1434.25 Tuesday 3 5 12 2.486168346815833E-8 2.01 0.08 id1148048 2 2016-06-04T14:22:55.000+0000 2016-06-05T14:19:45.000+0000 2 -74.00236 40.739872 -74.00156 40.746593 N 1436.83 Saturday 7 6 14 3.4685939141323567E-9 0.75 0.03 id2500786 2 2016-05-09T18:14:42.000+0000 2016-05-10T17:44:54.000+0000 1 -73.863625 40.769665 -73.983864 40.77387 N 1410.2 Monday 2 5 18 6.328009391024547E-7 10.14 0.43 id0843794 2 2016-01-08T08:09:48.000+0000 2016-01-09T07:23:50.000+0000 2 -73.99572 40.74894 -73.9799 40.752274 N 1394.03 Friday 6 1 8 1.1789215051701225E-8 1.38 0.06 id3375271 2 2016-05-20T11:38:47.000+0000 2016-05-21T11:24:31.000+0000 2 -73.96982 40.79725 -73.94046 40.81957 N 1425.73 Friday 6 5 11 7.554092952457263E-8 3.5 0.15 id0614225 2 2016-03-25T20:01:22.000+0000 2016-03-26T19:39:59.000+0000 2 -74.00107 40.720123 -73.98219 40.757572 N 1418.62 Friday 6 3 20 1.2237643089376416E-7 4.46 0.19 id0222074 2 2016-06-26T11:55:51.000+0000 2016-06-27T11:54:26.000+0000 3 -74.0153 40.704903 -73.98976 40.75736 N 1438.58 Sunday 1 6 11 2.3806288996551105E-7 6.22 0.26 id2891888 2 2016-06-05T10:26:37.000+0000 2016-06-06T10:18:51.000+0000 4 -73.95581 40.782352 -73.789894 40.64325 N 1432.23 Sunday 1 6 10 2.6780319825251107E-6 20.85 0.87 id2647716 2 2016-03-26T22:27:58.000+0000 2016-03-27T21:36:20.000+0000 1 -74.00077 40.74229 -73.9825 40.739548 N 1388.37 Saturday 7 3 22 1.516924322491745E-8 1.57 0.07 id2538452 2 2016-04-24T01:51:06.000+0000 2016-04-25T01:40:44.000+0000 1 -73.990906 40.760845 -73.99816 40.745148 N 1429.63 Sunday 1 4 1 2.1065835408621815E-8 1.85 0.08 id2090625 2 2016-02-16T15:01:25.000+0000 2016-02-17T14:31:41.000+0000 1 -73.96434 40.80775 -73.95491 40.803513 N 1410.27 Tuesday 3 2 15 5.247797220112619E-9 0.92 0.04 id0463051 2 2016-02-02T01:45:11.000+0000 2016-02-03T00:00:00.000+0000 1 -73.87084 40.773754 -73.971695 40.748768 N 1334.82 Tuesday 3 2 1 4.919345330197444E-7 8.94 0.4 id0587193 2 2016-05-23T18:01:22.000+0000 2016-05-24T17:56:46.000+0000 1 -73.99713 40.7254 -73.98323 40

In [46]:
%sql
select COUNT(*) from `train_csv` where trip_duration >= 900

count(1) 1973

1973 trips are there where trip duration is more than equal to 15hrs.

In [48]:
%sql
select COUNT(*) from `train_csv` where trip_duration >= 60

count(1) 12334

In [49]:
df.describe('speed').show()

+-------+------------------+
summary| speed|
+-------+------------------+
 count| 1458639|
 mean|14.422283258571843|
 stddev|14.975110069314914|
 min| 0.0|
 max| 9272.57|
+-------+------------------+

In [50]:
%sql
select speed from `train_csv`

speed 11.87 9.83 10.83 12.5 9.85 8.94 14.04 13.25 18.49 15.05 10.77 12.03 6.01 13.71 16.24 11.26 5.33 12.51 12.6 18.65 18.59 9.5 17.62 15.07 29.84 12.8 19.31 19.4 10.17 15.35 11.1 18.62 15.85 16.25 7.28 10.2 7.42 12.48 10.02 6.12 13.61 7.32 5.27 11.4 14.17 19.9 13.55 16.99 15.0 14.3 40.09 10.38 11.53 11.34 42.29 3.93 4.57 6.3 13.05 30.66 21.42 10.19 11.88 16.6 25.24 8.28 11.28 47.05 10.38 10.76 25.78 7.63 20.84 6.54 17.74 15.69 11.87 7.56 22.54 19.63 15.39 6.24 8.72 15.19 12.81 14.2 37.92 4.6 9.75 12.18 14.46 12.72 20.39 15.81 12.77 13.41 11.75 9.71 20.41 6.13 5.48 5.59 34.62 16.65 37.43 8.59 21.14 4.45 14.52 11.77 9.84 18.41 9.22 36.0 13.28 46.21 15.02 3.93 9.95 9.88 13.86 11.88 29.94 21.75 10.36 20.47 11.31 3.29 5.76 11.99 23.44 12.78 4.09 5.91 14.31 25.69 17.01 22.22 21.88 18.38 16.0 17.78 3.57 14.17 15.24 8.19 12.91 10.94 9.64 40.34 16.66 8.28 12.0 25.58 20.08 23.63 6.98 5.2 9.53 22.64 3.89 11.41 17.31 11.54 16.17 22.56 5.35 18.87 34.17 13.83 24.77 19.61 6.28 16.0 9.57 9.09 12.68 7.67 12.36 12.26 17.95 11.37 24.6 15.4 8.9 27.99 10.2 27.31 9.75 9.66 25.06 11.92 8.57 6.19 31.8 9.28 7.48 12.13 13.6 35.58 14.47 25.45 7.86 11.12 9.3 11.64 6.08 40.06 10.35 19.26 9.9 21.99 16.06 26.51 21.02 25.84 18.29 6.55 15.48 8.09 13.08 31.73 10.94 10.96 5.02 10.54 22.7 9.27 38.63 11.38 16.63 42.02 19.02 27.22 14.78 8.95 7.93 10.32 37.2 9.61 3.34 15.87 28.03 7.48 8.94 14.22 0.0 5.49 13.22 5.61 7.76 0.95 10.1 18.92 17.72 19.27 27.86 18.49 10.07 6.03 15.89 24.96 14.54 16.33 9.97 11.02 19.78 16.19 26.46 5.54 7.42 16.47 18.42 12.6 7.39 6.23 24.26 20.56 20.36 9.73 26.29 12.58 18.17 15.39 7.02 21.42 12.25 17.98 15.42 14.87 11.39 0.0 16.72 12.32 19.41 9.18 13.17 8.95 12.59 12.28 18.01 22.53 17.51 8.93 10.6 11.49 15.94 18.45 8.71 10.28 3.01 7.54 8.85 24.65 9.31 11.09 3.1 22.5 7.05 29.07 27.74 22.45 15.79 0.0 17.88 18.02 4.29 19.55 20.35 18.7 21.33 12.91 18.69 18.12 13.14 12.78 5.81 0.1 5.17 10.78 10.49 20.22 6.88 13.79 9.38 7.4 12.83 8.39 10.12 9.96 13.12 22.34 22.06 8.55 12.49 10.4 30.99 7.44 8.6 11.55 15.08 12.1 19.13 9.62 17.89 6.58 18.53 16.47 7.99 8.19 29.45 6.83 6.54 19.54 7.65 27.07 11.65 8.85 11.03 25.03 8.73 10.1 11.52 16.33 11.85 13.41 11.08 15.09 19.66 19.54 17.09 1.91 11.77 7.02 15.55 13.07 19.85 5.95 7.83 26.4 11.07 10.58 36.86 15.26 26.95 15.75 3.28 0.0 15.15 8.41 9.92 9.12 12.97 23.49 7.33 21.09 19.91 22.21 4.93 11.37 13.25 11.58 33.0 21.92 8.37 5.08 14.58 11.89 8.88 14.71 8.34 8.57 9.06 6.22 11.21 9.36 8.49 13.56 22.81 15.49 13.68 19.53 6.58 13.67 7.05 15.16 20.38 7.56 9.71 21.04 9.69 7.91 7.43 8.28 12.66 14.79 11.5 8.45 13.53 20.66 10.87 28.24 18.55 6.68 29.81 7.82 36.89 11.45 17.72 17.16 7.61 30.42 10.51 27.73 8.75 15.66 5.63 14.55 19.41 22.99 6.71 19.18 8.5 14.9 9.78 28.49 6.27 6.76 4.77 12.27 7.71 7.9 18.5 30.07 14.14 12.67 11.75 8.75 13.24 15.04 9.23 12.72 13.63 8.89 21.1 5.37 11.63 4.66 10.07 7.16 31.21 6.42 9.46 5.06 8.46 34.07 10.51 24.86 12.37 16.75 5.3 12.88 11.22 6.92 11.25 18.47 13.75 7.18 12.73 18.64 13.66 0.13 9.99 6.29 10.84 18.49 17.13 13.99 8.0 13.3 6.63 5.98 20.82 33.3 27.44 23.08 12.73 20.26 5.39 12.04 5.96 32.6 0.08 9.56 5.96 12.28 13.92 15.55 27.9 11.61 7.79 5.44 13.35 6.84 5.92 6.53 17.71 12.1 17.26 12.45 14.26 13.29 9.35 9.02 19.85 9.02 16.13 13.85 4.06 16.86 15.97 5.34 5.09 13.07 11.6 5.82 14.53 8.78 13.79 7.67 24.86 3.08 25.71 19.65 11.83 11.54 14.85 6.09 17.9 27.98 14.25 7.3 9.28 12.64 6.01 13.41 12.91 41.71 13.13 11.82 11.08 7.14 20.08 17.56 9.65 19.02 13.25 9.19 17.21 14.4 31.79 7.16 5.64 12.78 18.46 8.72 8.88 15.01 14.22 20.6 18.57 24.0 15.07 19.45 32.93 8.34 8.7 44.72 15.95 10.01 24.25 11.33 32.53 21.24 13.54 15.16 19.17 17.94 16.24 7.21 13.41 13.76 12.03 33.57 7.59 14.16 18.0 8.22 21.22 5.69 9.8 10.26 27.0 9.96 7.05 7.95 4.09 28.13 9.21 8.53 21.22 3.08 6.4 16.98 14.02 17.08 11.45 17.93 13.24 7.52 5.17 15.14 21.87 28.69 7.35 10.27 7.03 11.65 10.66 27.78 25.12 23.11 39.63 10.08 11.86 8.18 26.36 14.97 7.53 3.02 13.86 14.99 0.0 14.79 7.31 11.34 24.35 17.14 21.47 14.83 20.05 27.43 20.77 10.76 23.3 14.8 4.8 20

Most of time car speed was 9 to 18, as because it is in city area.
There are speed more than 60, may be it is because outside city and long tour, next investigate that

In [52]:
%sql
select * from `train_csv` where trip_duration >= 900 and speed >= 50

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration weekday weekday_num month pickup_hour a distance speed

In [53]:
%sql
CREATE TABLE train_csv_delta
USING delta FROM `train_csv`

In [54]:
%sql
DROP table train_csv_delta

In [55]:
%sql
select MAX(speed) from `train_csv` where trip_duration >= 800 

max(speed) 2.55

We can easily verify the trips which takes more than 800 mins is false trip because the average speed is 2.55 km  So we can delete these trip

In [57]:
df = df.filter(col("trip_duration") < 800)

In [58]:
df.createOrReplaceTempView(temp_table_name)

In [59]:
%sql
select COUNT(speed) from `train_csv` where trip_duration >= 800 

count(speed) 0

In [60]:
%sql
select COUNT(speed) from `train_csv` where speed >= 120 

count(speed) 135

We also can delete superfast taxi, speed more than 100

In [62]:
df = df.filter(col("speed") <= 100)

In [63]:
df.createOrReplaceTempView(temp_table_name)

In [64]:
%sql
select COUNT(speed) from `train_csv` where speed > 100 

count(speed) 0

In [65]:
df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: timestamp (nullable = true)
-- dropoff_datetime: timestamp (nullable = true)
-- passenger_count: string (nullable = true)
-- pickup_longitude: float (nullable = true)
-- pickup_latitude: float (nullable = true)
-- dropoff_longitude: float (nullable = true)
-- dropoff_latitude: float (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: double (nullable = true)
-- weekday: string (nullable = true)
-- weekday_num: string (nullable = true)
-- month: string (nullable = true)
-- pickup_hour: string (nullable = true)
-- a: double (nullable = true)
-- distance: double (nullable = true)
-- speed: double (nullable = true)

In [66]:
%sql
select distinct store_and_fwd_flag, count(store_and_fwd_flag) as Count from `train_csv` group by store_and_fwd_flag order by store_and_fwd_flag

store_and_fwd_flag Count N 1448441 Y 8044

In [67]:
%sql
select distinct pickup_hour, count(pickup_hour) as Count from `train_csv` group by pickup_hour order by Count

pickup_hour Count 5 14977 4 15752 3 20842 2 27906 6 33213 1 38491 0 53160 7 55546 16 64214 10 65366 8 66976 9 67582 11 68396 23 69683 13 71357 15 71677 12 71772 14 74174 17 76389 22 80345 20 83954 21 84063 19 90190 18 90460

In [68]:
%sql
select distinct weekday, count(weekday) as Count from `train_csv` group by weekday order by Count

weekday Count Monday 187198 Sunday 195014 Tuesday 202479 Wednesday 209872 Thursday 218267 Saturday 220467 Friday 223188

In [69]:
%sql
select distinct month, count(month) as TripCounts from `train_csv` group by month order by TripCounts

month TripCounts 1 229360 6 233942 2 237977 5 248119 4 251277 3 255810

In [70]:
%sql
select distinct pickup_hour, avg(trip_duration) as AvgDuration from `train_csv` group by pickup_hour

pickup_hour AvgDuration 7 12.648269722392254 15 16.212865354297758 11 14.669792824141798 3 11.739814317244024 8 13.912657220496898 22 13.415201941626737 16 16.1709882580123 0 13.021907072987208 5 11.872748213928025 18 14.378360048640273 17 15.578413645943773 6 11.196112666726881 19 13.168667036256764 23 13.54081382833689 9 14.068184723743002 1 12.335211607908342 20 12.84204409557615 10 14.14935516935409 4 12.234386744540382 12 14.693859304464132 13 15.073300026626667 14 15.863650335697162 21 12.95259638604381 2 11.689348885544337

In [71]:
%sql
select distinct weekday_num, avg(trip_duration) as AvgDuration from `train_csv` group by weekday_num order by weekday_num

weekday_num AvgDuration 1 12.75195201370157 2 13.54589418690373 3 14.300533833138282 4 14.672807330182222 5 14.978421978585851 6 14.487531632525037 7 13.006548644468344

In [72]:
%sql
select distinct month, avg(trip_duration) as AvgDuration from `train_csv` group by month order by month

month AvgDuration 1 13.210178365887684 2 13.19360438193608 3 13.670138188499285 4 14.16830772414506 5 14.7588331405495 6 14.871464508296919

In [73]:
%sql
select distinct trip_duration, distance  from `train_csv`

trip_duration distance 8.78 2.76 11.72 1.97 4.37 0.84 9.42 1.91 4.97 1.2 8.32 0.69 4.93 1.0 45.37 8.72 9.32 2.95 12.87 1.67 6.5 1.16 7.3 1.23 11.42 1.36 9.22 1.84 12.07 3.26 9.63 1.29 8.3 3.99 15.77 1.03 16.68 3.15 2.38 1.06 8.75 2.49 35.45 19.33 7.38 1.43 11.3 4.49 7.68 2.09 9.97 1.82 5.03 1.09 16.9 3.7 32.95 9.44 40.97 3.27 28.08 5.94 11.53 3.03 8.53 2.76 12.82 4.66 22.62 12.77 22.25 11.95 13.93 4.53 6.97 2.02 2.55 0.93 9.9 0.71 20.37 7.27 11.03 0.88 4.12 0.89 2.28 0.71 1.78 0.53 2.25 0.49 15.18 3.57 13.35 4.57 5.43 1.35 19.32 5.17 10.08 1.37 8.45 1.45 29.4 6.8 18.42 2.74 9.33 1.33 9.15 2.99 6.4 1.36 5.85 1.02 10.68 3.28 4.62 0.61 24.1 6.37 7.3 2.18 20.68 3.25 11.13 1.02 4.12 1.07 19.3 7.25 8.6 1.76 17.97 6.0 7.67 1.21 15.35 6.07 15.25 3.44 23.18 2.74 3.7 1.43 13.83 1.74 68.47 10.78 7.97 1.16 1.95 0.38 11.28 0.0 7.3 2.14 8.37 1.64 44.45 0.92 8.72 1.29 4.53 0.41 10.18 0.75 11.65 6.75 14.15 4.58 8.95 2.22 5.65 1.58 22.17 6.77 8.78 1.4 11.65 3.38 25.38 18.98 24.18 2.61 11.47 1.06 14.53 1.9 49.8 6.33 4.43 1.13 2.15 0.76 26.5 8.9 8.8 0.89 29.4 3.58 12.93 2.32 18.1 3.89 12.0 3.51 21.6 3.31 6.03 1.22 5.12 1.19 11.97 2.01 23.32 5.48 18.87 11.21 14.88 3.43 17.48 12.15 20.28 2.58 47.55 9.45 65.65 20.72 15.27 1.45 20.07 5.45 15.15 2.72 7.33 1.57 6.73 2.23 10.6 1.72 4.32 1.64 3.63 0.7 13.53 3.05 6.7 1.01 15.38 4.54 18.8 5.82 1.52 1.0 15.38 3.52 13.47 5.51 28.85 8.44 16.97 2.72 10.48 1.51 20.68 2.33 6.47 1.28 15.72 1.58 13.85 1.56 5.77 1.06 6.83 0.86 11.78 1.33 8.67 2.84 52.17 13.25 22.27 3.57 8.88 1.37 4.5 1.33 6.4 1.25 27.93 1.86 19.02 7.05 2.28 0.5 19.3 3.62 4.53 2.21 20.28 7.93 7.57 2.0 9.03 3.13 6.9 1.68 9.77 0.64 6.5 1.57 13.92 3.24 12.65 3.32 21.75 2.35 8.42 0.95 28.42 3.11 26.55 15.68 3.95 0.93 4.37 3.35 17.13 2.41 9.65 1.11 7.83 2.06 17.78 3.43 8.55 1.99 18.28 4.45 8.5 3.09 44.42 4.83 9.57 1.8 28.0 13.77 18.37 2.18 2.5 0.56 5.73 1.0 15.98 4.84 7.57 0.63 11.7 3.77 24.98 6.2 22.22 4.53 6.55 0.98 19.4 4.72 37.03 13.33 37.7 14.31 7.38 0.92 19.78 3.05 18.72 2.86 24.17 4.26 43.05 8.51 5.17 1.3 10.45 1.27 21.37 8.12 12.7 2.55 6.32 0.76 13.78 8.69 14.87 2.56 13.15 2.27 20.68 1.71 14.67 1.11 4.65 1.54 19.32 8.18 10.45 2.01 24.32 2.47 20.38 6.68 21.67 9.03 26.93 5.06 13.88 6.99 16.38 2.0 17.73 6.83 14.7 3.06 17.8 4.8 11.98 2.0 5.63 1.8 3.07 0.77 12.63 2.91 15.87 5.97 9.3 0.68 9.75 1.99 11.18 2.89 24.52 2.67 20.68 3.0 8.82 3.17 18.13 2.74 41.7 16.05 8.4 2.1 12.97 3.76 13.55 2.5 20.23 6.0 14.15 1.96 5.42 0.53 14.38 3.66 3.18 0.54 34.1 9.48 17.0 4.28 22.27 7.12 19.77 6.25 9.68 2.81 18.82 6.54 10.15 2.53 7.3 1.46 6.05 3.1 4.95 0.77 8.72 2.01 6.5 1.31 24.58 1.96 10.62 1.2 16.08 2.93 22.07 3.48 13.53 3.61 17.02 6.51 2.83 0.48 6.2 1.76 19.92 4.83 15.8 4.91 32.43 2.64 10.13 1.45 12.95 3.77 17.35 3.36 10.63 3.18 25.35 4.03 6.38 2.04 8.15 1.38 13.83 6.32 15.35 3.13 20.98 1.53 4.35 0.54 0.82 0.1 13.75 1.78 12.8 3.45 21.68 2.47 15.45 3.1 8.98 1.84 12.33 3.14 40.03 9.33 9.53 2.03 23.27 7.44 7.9 1.89 23.87 7.28 7.93 2.75 22.9 6.62 13.6 3.24 5.6 1.57 19.58 7.83 33.38 10.3 4.82 1.23 11.4 2.88 29.0 1.84 16.72 4.02 12.45 2.96 8.35 1.41 7.83 2.59 2.73 0.87 7.8 1.49 13.87 1.83 11.63 2.08 22.98 4.11 28.13 9.53 23.25 2.28 13.82 4.06 4.9 0.75 21.18 12.27 23.1 9.46 6.4 1.46 10.97 1.49 13.35 2.4 24.67 8.63 13.85 7.99 16.9 2.71 2.07 0.43 56.82 6.14 2.62 0.75 6.33 1.74 18.13 1.41 9.2 4.33 6.25 1.16 15.3 0.9 14.03 5.8 13.02 3.69 5.92 1.76 38.27 13.81 40.13 7.52 11.38 3.58 16.55 2.94 10.63 2.26 7.27 1.87 6.52 1.73 8.37 0.99 14.7 4.42 24.27 14.28 16.27 1.93 9.37 6.16 11.93 1.2 1.58 0.95 27.5 20.19 2.05 0.76 6.47 1.96 17.68 5.18 13.78 4.3 5.55 1.75 19.22 2.99 8.0 0.76 3.83 0.72 34.67 8.01 18.15 3.15 20.25 8.45 10.73 1.07 23.85 5.81 7.27 0.32 4.02 1.24 18.18 2.01 13.0 3.16 7.55 1.88 9.27 3.25 20.97 6.15 21.5 3.8 10.55 1.87 20.42 3.33 3.92 1.47 38.37 4.73 19.68 7.44 29.87 3.14 12.33 5.8 16.55 2.74 9.83 3.96 19.42 2.85 12.73 3.77 2.9 0.35 16.27 2.98 7.3 0.72 16.2 1.99 4.37 1.05 36.7 9.09 6.03 2.12 21.93 4.34 9.08 1.94 4.1 0.58 11.73 1.91 8.83 2.69 5.65 0.81 5.65 0.63 18.7

In [74]:
%sql
select distinct pickup_longitude, pickup_latitude  from `train_csv`

pickup_longitude pickup_latitude -73.989815 40.72965 -73.9913 40.71707 -73.98217 40.75572 -74.01551 40.71584 -73.974686 40.783226 -74.00416 40.71297 -74.0091 40.70707 -73.99668 40.72586 -73.97695 40.764442 -73.99577 40.725147 -73.98455 40.745537 -73.98968 40.76057 -73.96465 40.80715 -74.00728 40.74347 -73.97835 40.78827 -74.00174 40.73563 -73.93849 40.84624 -74.002014 40.73954 -73.7894 40.646633 -73.96921 40.800808 -73.98852 40.72333 -73.963524 40.75748 -73.952675 40.776577 -73.977684 40.74626 -73.95719 40.779804 -73.87446 40.773945 -73.98624 40.76008 -73.99409 40.758865 -73.99783 40.72623 -73.99904 40.734283 -73.96154 40.76865 -73.87414 40.77372 -73.95659 40.675213 -73.94986 40.787216 -73.97335 40.74836 -73.97894 40.75838 -73.97675 40.74428 -73.9928 40.68981 -73.901024 40.741306 -73.96755 40.756588 -73.95332 40.778534 -74.00854 40.70802 -74.00399 40.74785 -74.00507 40.71921 -73.98694 40.740494 -73.95436 40.76399 -73.994675 40.740196 -73.9918 40.76485 -73.98721 40.75622 -73.98923 40.74308 -73.96943 40.76383 -73.98735 40.73879 -73.95654 40.7722 -73.95456 40.686363 -73.9595 40.767403 -74.00896 40.741463 -73.98639 40.75687 -74.00631 40.73339 -73.9536 40.766937 -73.99656 40.7609 -73.7769 40.64511 -73.97654 40.750793 -74.003975 40.707565 -73.952354 40.77731 -73.9921 40.750053 -73.96371 40.771294 -73.99199 40.73779 -73.78187 40.644764 -73.99239 40.749317 -73.98716 40.767044 -73.98488 40.696995 -73.78999 40.64354 -73.8857 40.77315 -73.99417 40.751156 -73.98755 40.71971 -73.96266 40.76682 -73.967834 40.76839 -73.976715 40.7661 -73.89639 40.74687 -73.99403 40.724365 -74.001945 40.739624 -73.99434 40.74025 -73.97047 40.78414 -73.99046 40.75096 -73.99289 40.74779 -73.99071 40.74163 -73.98878 40.72216 -73.97478 40.758717 -73.96974 40.75757 -73.97226 40.796574 -73.98334 40.76598 -73.99974 40.76101 -73.96377 40.79804 -73.991844 40.74952 -73.98552 40.74705 -73.98717 40.733078 -73.97715 40.76453 -74.00269 40.760334 -73.96566 40.76226 -73.99132 40.72996 -73.975975 40.79147 -73.86413 40.770203 -73.95911 40.759613 -73.99175 40.713993 -73.97179 40.79461 -73.98292 40.731323 -73.7883 40.64135 -73.9748 40.78809 -73.98122 40.75837 -74.00957 40.706028 -73.94807 40.71766 -73.97804 40.762447 -73.95765 40.76164 -73.991684 40.74544 -74.00634 40.73317 -73.77674 40.64521 -73.98455 40.76968 -73.979355 40.765167 -73.98469 40.769093 -73.98541 40.75257 -73.99017 40.759193 -73.94878 40.776627 -73.93944 40.70562 -73.98682 40.761295 -73.9602 40.77301 -73.97056 40.76138 -73.9876 40.732754 -73.98243 40.77725 -73.99979 40.73322 -74.00275 40.73929 -73.98827 40.759377 -73.97193 40.747253 -73.98141 40.757362 -73.9838 40.72565 -73.97652 40.76411 -73.99124 40.749714 -74.00169 40.74554 -74.00241 40.75553 -73.95291 40.786747 -73.977776 40.746117 -73.96655 40.78919 -73.96207 40.759457 -73.98049 40.78316 -73.94494 40.79412 -73.9378 40.84836 -73.938385 40.79618 -74.0025 40.739628 -73.99167 40.75631 -73.98593 40.74021 -73.98053 40.75962 -73.99005 40.75198 -73.95397 40.77497 -73.99402 40.75219 -73.979416 40.76346 -73.96404 40.71322 -73.96803 40.802296 -74.00998 40.712147 -74.00991 40.720833 -73.98372 40.74085 -73.98557 40.759693 -73.99971 40.67399 -73.971245 40.751305 -74.0136 40.71384 -73.954025 40.781746 -73.98607 40.76756 -73.99118 40.729828 -74.0011 40.75715 -73.991264 40.755695 -73.78191 40.64479 -73.98891 40.730797 -73.96474 40.801895 -73.97009 40.76499 -73.97701 40.75335 -73.97826 40.783283 -74.00021 40.73285 -73.86329 40.76986 -73.9692 40.754475 -74.00976 40.720413 -73.99266 40.723927 -73.78463 40.64861 -73.92624 40.70334 -74.00884 40.715935 -73.95781 40.801105 -73.97383 40.75683 -73.992325 40.72495 -74.000496 40.7372 -74.0151 40.718327 -73.96564 40.75857 -73.97425 40.746956 -73.97727 40.68126 -73.9794 40.74565 -73.98192 40.756596 -73.97086 40.75563 -73.98226 40.74571 -73.9695 40.78545 -74.00067 40.714397 -73.97156 40.786194 -74.00461 40.73071 -73.996185 40.727325 -73.99684 40.71473 -73.969604 40.760372 -73.987404 40.744114 -73.99133 40.766487 -73.96745 40.75677 -74.0053

In [75]:
%sql
select distinct dropoff_longitude, dropoff_latitude  from `train_csv`

dropoff_longitude dropoff_latitude -73.99624 40.72537 -73.963104 40.768974 -73.98538 40.69084 -74.00432 40.73326 -73.98217 40.769073 -73.94968 40.713264 -74.0048 40.70965 -73.99486 40.726948 -73.99143 40.754784 -74.003624 40.717438 -74.00635 40.706566 -73.954834 40.692814 -73.99188 40.74436 -73.979 40.753323 -73.97383 40.737743 -73.9906 40.75782 -73.92888 40.67054 -73.96231 40.77816 -73.98316 40.76202 -73.99004 40.71403 -73.92967 40.85478 -73.88294 40.74701 -73.9888 40.734768 -73.99348 40.74291 -73.97777 40.754417 -73.982895 40.76021 -73.97755 40.73888 -73.97699 40.78002 -73.978516 40.75523 -73.97624 40.79112 -73.957115 40.77795 -73.94015 40.82547 -73.99452 40.75035 -73.96407 40.798077 -73.98253 40.754925 -73.95606 40.78456 -74.00135 40.75639 -73.98642 40.718273 -73.833984 40.758327 -73.997284 40.721687 -74.00467 40.741753 -74.00891 40.730434 -73.99183 40.759 -73.90043 40.86245 -73.9874 40.753468 -73.95405 40.766777 -73.96656 40.792496 -73.77681 40.64327 -73.89972 40.74017 -73.96385 40.770237 -73.984505 40.73224 -73.977356 40.738686 -73.97811 40.76146 -73.97829 40.68459 -74.00482 40.74695 -73.951805 40.773544 -73.97264 40.78098 -74.0132 40.70926 -73.99537 40.721622 -74.01311 40.715305 -73.971436 40.79526 -73.95274 40.786007 -73.9742 40.75721 -73.97068 40.8001 -73.97567 40.741432 -73.96456 40.755882 -73.95554 40.77935 -73.95039 40.77537 -74.00326 40.74375 -73.98124 40.758934 -73.96511 40.672512 -73.945305 40.78849 -73.98072 40.75056 -73.991005 40.75073 -73.975204 40.76526 -73.99673 40.741497 -73.95376 40.76469 -73.96955 40.757668 -73.98916 40.72597 -73.971825 40.76247 -73.992615 40.758488 -73.97817 40.778027 -73.9776 40.74935 -73.971695 40.743626 -73.95528 40.77302 -74.00876 40.728184 -73.9773 40.75843 -73.96347 40.77447 -73.98959 40.75927 -73.9908 40.724327 -73.95775 40.7728 -73.954735 40.769657 -74.002655 40.76063 -73.99159 40.75017 -73.960236 40.763615 -73.97948 40.74963 -73.99122 40.687805 -73.99229 40.73369 -73.98557 40.75949 -73.97968 40.76336 -74.00281 40.74237 -74.00758 40.740803 -73.9854 40.77855 -73.98489 40.76211 -73.976105 40.64553 -73.99065 40.756058 -73.96412 40.77033 -73.99223 40.75184 -73.99967 40.71391 -73.98307 40.754345 -73.97383 40.761375 -73.99054 40.71939 -73.984955 40.721207 -73.957855 40.77932 -73.97552 40.75627 -73.96047 40.77584 -74.00176 40.728046 -73.98457 40.73966 -73.99219 40.749146 -73.97968 40.76565 -73.98995 40.748543 -73.96143 40.756546 -73.96353 40.765507 -73.97541 40.78231 -73.990814 40.765728 -74.00426 40.75267 -73.991005 40.750378 -73.98502 40.7484 -73.95608 40.770123 -73.99919 40.738415 -73.97606 40.75138 -73.97724 40.76445 -73.9855 40.76313 -73.93529 40.797558 -73.99521 40.76014 -73.9836 40.755985 -73.97781 40.756912 -73.99565 40.73196 -73.9721 40.757465 -73.9787 40.762154 -73.90223 40.76297 -73.97425 40.7546 -73.99756 40.727264 -73.98617 40.767345 -73.973076 40.79204 -74.182465 40.687767 -73.97818 40.758125 -73.96109 40.76454 -73.97957 40.75293 -73.95296 40.785927 -73.98179 40.75575 -74.005554 40.726616 -73.96004 40.77657 -73.7898 40.646973 -73.990814 40.745365 -73.9999 40.7294 -74.00049 40.72886 -73.9858 40.75283 -74.00433 40.72187 -73.960335 40.77016 -73.96464 40.775806 -73.968704 40.751823 -73.92426 40.75609 -73.99145 40.73185 -74.00167 40.727955 -74.01694 40.70483 -73.942154 40.79889 -73.95644 40.747364 -73.9839 40.743847 -73.94679 40.779953 -73.95241 40.77806 -73.99409 40.75612 -73.96044 40.76295 -73.98854 40.74856 -74.002495 40.748695 -73.95597 40.78186 -73.9728 40.75857 -73.99132 40.749752 -73.99161 40.73044 -73.992035 40.74402 -73.975075 40.787457 -74.01757 40.711407 -73.97019 40.79661 -73.9664 40.681423 -73.96145 40.71638 -74.003876 40.72964 -73.97759 40.766087 -73.98098 40.784782 -74.008125 40.738995 -73.99591 40.74337 -74.0039 40.735542 -73.77659 40.64505 -73.96085 40.76114 -73.99425 40.755142 -73.99169 40.75812 -73.965645 40.76609 -73.98307 40.730686 -73.98312 40.731483 -73.978294 40.752056 -73.98116 40.75197 -74.01468 40.71405 -73.97368 40.757156 -73.9776 40.753796 -73.98270

In [76]:
df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: timestamp (nullable = true)
-- dropoff_datetime: timestamp (nullable = true)
-- passenger_count: string (nullable = true)
-- pickup_longitude: float (nullable = true)
-- pickup_latitude: float (nullable = true)
-- dropoff_longitude: float (nullable = true)
-- dropoff_latitude: float (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: double (nullable = true)
-- weekday: string (nullable = true)
-- weekday_num: string (nullable = true)
-- month: string (nullable = true)
-- pickup_hour: string (nullable = true)
-- a: double (nullable = true)
-- distance: double (nullable = true)
-- speed: double (nullable = true)

In [77]:
df = df.withColumn("avgSpeed", lit(16))

In [78]:
df.printSchema()

root
-- id: string (nullable = true)
-- vendor_id: string (nullable = true)
-- pickup_datetime: timestamp (nullable = true)
-- dropoff_datetime: timestamp (nullable = true)
-- passenger_count: string (nullable = true)
-- pickup_longitude: float (nullable = true)
-- pickup_latitude: float (nullable = true)
-- dropoff_longitude: float (nullable = true)
-- dropoff_latitude: float (nullable = true)
-- store_and_fwd_flag: string (nullable = true)
-- trip_duration: double (nullable = true)
-- weekday: string (nullable = true)
-- weekday_num: string (nullable = true)
-- month: string (nullable = true)
-- pickup_hour: string (nullable = true)
-- a: double (nullable = true)
-- distance: double (nullable = true)
-- speed: double (nullable = true)
-- avgSpeed: integer (nullable = false)

In [79]:
display(df)

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration weekday weekday_num month pickup_hour a distance speed avgSpeed id2875421 2 2016-03-14T17:24:55.000+0000 2016-03-14T17:32:30.000+0000 1 -73.982155 40.767937 -73.96463 40.765602 N 7.58 Monday 2 3 17 1.3830896636179652E-8 1.5 11.87 16 id2377394 1 2016-06-12T00:43:35.000+0000 2016-06-12T00:54:38.000+0000 1 -73.980415 40.738564 -73.99948 40.73115 N 11.05 Sunday 1 6 0 2.0078128522717423E-8 1.81 9.83 16 id3858529 2 2016-01-19T11:35:24.000+0000 2016-01-19T12:10:48.000+0000 1 -73.97903 40.76394 -74.00533 40.710087 N 35.4 Tuesday 3 1 11 2.511076618143245E-7 6.39 10.83 16 id3504673 2 2016-04-06T19:32:31.000+0000 2016-04-06T19:39:40.000+0000 1 -74.01004 40.71997 -74.01227 40.70672 N 7.15 Wednesday 4 4 19 1.359155624250055E-8 1.49 12.5 16 id2181028 2 2016-03-26T13:30:55.000+0000 2016-03-26T13:38:10.000+0000 1 -73.97305 40.79321 -73.97292 40.78252 N 7.25 Saturday 7 3 13 8.701373594095704E-9 1.19 9.85 16 id0801584 2 2016-01-30T22:01:40.000+0000 2016-01-30T22:09:03.000+0000 6 -73.98286 40.742195 -73.99208 40.749184 N 7.38 Saturday 7 1 22 7.438317364467243E-9 1.1 8.94 16 id1813257 1 2016-06-17T22:34:59.000+0000 2016-06-17T22:40:40.000+0000 4 -73.96902 40.75784 -73.957405 40.765896 N 5.68 Friday 6 6 22 1.0834136581882085E-8 1.33 14.04 16 id1324603 2 2016-05-21T07:54:58.000+0000 2016-05-21T08:20:49.000+0000 1 -73.96928 40.79778 -73.92247 40.76056 N 25.85 Saturday 7 5 7 2.0116586779003735E-7 5.71 13.25 16 id1301050 1 2016-05-27T23:12:23.000+0000 2016-05-27T23:16:38.000+0000 1 -73.99948 40.7384 -73.98579 40.732815 N 4.25 Friday 6 5 23 1.0575516792025634E-8 1.31 18.49 16 id0012891 2 2016-03-10T21:45:01.000+0000 2016-03-10T22:05:26.000+0000 1 -73.98105 40.74434 -73.973 40.78999 N 20.42 Thursday 5 3 21 1.6153317501260625E-7 5.12 15.05 16 id1436371 2 2016-05-10T22:08:41.000+0000 2016-05-10T22:29:55.000+0000 1 -73.98265 40.76384 -74.00223 40.73299 N 21.23 Tuesday 3 5 22 8.922656314182737E-8 3.81 10.77 16 id1299289 2 2016-05-15T11:16:11.000+0000 2016-05-15T11:34:59.000+0000 4 -73.99153 40.74944 -73.95654 40.77063 N 18.8 Sunday 1 5 11 8.768402598600243E-8 3.77 12.03 16 id1187965 2 2016-02-19T09:52:46.000+0000 2016-02-19T10:11:20.000+0000 2 -73.96298 40.75668 -73.984406 40.76072 N 18.57 Friday 6 2 9 2.129654877938154E-8 1.86 6.01 16 id0799785 2 2016-06-01T20:58:29.000+0000 2016-06-01T21:02:49.000+0000 1 -73.95631 40.76794 -73.96611 40.763 N 4.33 Wednesday 4 6 20 6.057203368798397E-9 0.99 13.71 16 id2900608 2 2016-05-27T00:43:36.000+0000 2016-05-27T01:07:10.000+0000 1 -73.992195 40.727226 -73.974655 40.78307 N 23.57 Friday 6 5 0 2.509297192595758E-7 6.38 16.24 16 id3319787 1 2016-05-16T15:29:02.000+0000 2016-05-16T15:32:33.000+0000 1 -73.95551 40.768593 -73.94876 40.771545 N 3.52 Monday 2 5 15 2.655201508351016E-9 0.66 11.26 16 id3379579 2 2016-04-11T17:29:50.000+0000 2016-04-11T18:08:26.000+0000 1 -73.991165 40.75556 -73.99929 40.725353 N 38.6 Monday 2 4 17 7.238160272905697E-8 3.43 5.33 16 id1154431 1 2016-04-14T08:48:26.000+0000 2016-04-14T09:00:37.000+0000 1 -73.994255 40.745804 -73.99966 40.723343 N 12.18 Thursday 5 4 8 3.9695192884170783E-8 2.54 12.51 16 id3552682 1 2016-06-27T09:55:13.000+0000 2016-06-27T10:17:10.000+0000 1 -74.00398 40.713013 -73.979195 40.749924 N 21.95 Monday 2 6 9 1.3062369794697426E-7 4.61 12.6 16 id3390316 2 2016-06-05T13:47:23.000+0000 2016-06-05T13:51:34.000+0000 1 -73.98389 40.738197 -73.9912 40.72787 N 4.18 Sunday 1 6 13 1.0461510732205175E-8 1.3 18.65 16 id2070428 1 2016-02-28T02:23:02.000+0000 2016-02-28T02:31:08.000+0000 1 -73.98037 40.74242 -73.96285 40.760635 N 8.1 Sunday 1 2 2 3.867776267452501E-8 2.51 18.59 16 id0809232 2 2016-04-01T12:12:25.000+0000 2016-04-01T12:23:17.000+0000 1 -73.97954 40.75336 -73.964 40.76346 N 10.87 Friday 6 4 12 1.8317931467470727E-8 1.72 9.5 16 id2352683 1 2016-04-09T03:34:27.000+0000 2016-04-09T03:41:30.000+0000 1 -73.995865 40

Convert categorical to one hot encoder

In [81]:
from pyspark.ml import Pipeline
from pyspark.ml.feature import OneHotEncoderEstimator, StringIndexer, VectorAssembler
categoricalColumns = ["passenger_count", "weekday_num", "month", "pickup_hour"]
stages = []
for categoricalCol in categoricalColumns:
    # Category Indexing with StringIndexer
    stringIndexer = StringIndexer(inputCol=categoricalCol, outputCol=categoricalCol + "Index")
    print(stringIndexer,type(stringIndexer))
    # Use OneHotEncoder to convert categorical variables into binary SparseVectors
    # encoder = OneHotEncoderEstimator(inputCol=categoricalCol + "Index", outputCol=categoricalCol + "classVec")
    encoder = OneHotEncoderEstimator(inputCols=[stringIndexer.getOutputCol()], outputCols=[categoricalCol + "classVec"])
    # Add stages.  These are not run here, but will run all at once later on.
    stages += [stringIndexer, encoder]

StringIndexer_b9902fc37593 <class 'pyspark.ml.feature.StringIndexer'>
StringIndexer_16e1243def50 <class 'pyspark.ml.feature.StringIndexer'>
StringIndexer_bcc6821bb742 <class 'pyspark.ml.feature.StringIndexer'>
StringIndexer_1adc0eccbe31 <class 'pyspark.ml.feature.StringIndexer'>

Use a VectorAssembler to combine all the feature columns into a single vector column

In [83]:
print(stages)

[StringIndexer_b9902fc37593, OneHotEncoderEstimator_2620ab545e6b, StringIndexer_16e1243def50, OneHotEncoderEstimator_74aecd803fda, StringIndexer_bcc6821bb742, OneHotEncoderEstimator_c05cc592dbad, StringIndexer_1adc0eccbe31, OneHotEncoderEstimator_eed4fd7b8970]

In [84]:
numericCols = ["pickup_longitude", "pickup_latitude", "dropoff_longitude", "dropoff_latitude", "distance","avgSpeed"]
assemblerInputs = [c + "classVec" for c in categoricalColumns] + numericCols
assembler = VectorAssembler(inputCols=assemblerInputs, outputCol="features")
stages += [assembler]

In [85]:
print(stages)

[StringIndexer_b9902fc37593, OneHotEncoderEstimator_2620ab545e6b, StringIndexer_16e1243def50, OneHotEncoderEstimator_74aecd803fda, StringIndexer_bcc6821bb742, OneHotEncoderEstimator_c05cc592dbad, StringIndexer_1adc0eccbe31, OneHotEncoderEstimator_eed4fd7b8970, VectorAssembler_311276a65ba0]

In [86]:
#df_stages =  assembler.transform(df)

In [87]:
#from pyspark.ml.classification import LogisticRegression
  
partialPipeline = Pipeline().setStages(stages)
pipelineModel = partialPipeline.fit(df)
preppedDataDF = pipelineModel.transform(df)

In [88]:
type(preppedDataDF)

Out[ 48 ]: pyspark.sql.dataframe.DataFrame

In [89]:
trip_duration = df.select("trip_duration").collect()

In [90]:
type(trip_duration)

Out[ 50 ]: list

In [91]:
preppedDataDF = preppedDataDF.withColumn("label", df.trip_duration)

In [92]:
display(preppedDataDF)

id vendor_id pickup_datetime dropoff_datetime passenger_count pickup_longitude pickup_latitude dropoff_longitude dropoff_latitude store_and_fwd_flag trip_duration weekday weekday_num month pickup_hour a distance speed avgSpeed passenger_countIndex passenger_countclassVec weekday_numIndex weekday_numclassVec monthIndex monthclassVec pickup_hourIndex pickup_hourclassVec features label id2875421 2 2016-03-14T17:24:55.000+0000 2016-03-14T17:32:30.000+0000 1 -73.982155 40.767937 -73.96463 40.765602 N 7.58 Monday 2 3 17 1.3830896636179652E-8 1.5 11.87 16 0.0 List(0, 5, List(0), List(1.0)) 6.0 List(0, 6, List(), List()) 0.0 List(0, 5, List(0), List(1.0)) 5.0 List(0, 23, List(5), List(1.0)) List(0, 45, List(0, 11, 21, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, -73.9821548461914, 40.76793670654297, -73.96463012695312, 40.765602111816406, 1.5, 16.0)) 7.58 id2377394 1 2016-06-12T00:43:35.000+0000 2016-06-12T00:54:38.000+0000 1 -73.980415 40.738564 -73.99948 40.73115 N 11.05 Sunday 1 6 0 2.0078128522717423E-8 1.81 9.83 16 0.0 List(0, 5, List(0), List(1.0)) 5.0 List(0, 6, List(5), List(1.0)) 4.0 List(0, 5, List(4), List(1.0)) 17.0 List(0, 23, List(17), List(1.0)) List(0, 45, List(0, 10, 15, 33, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, 1.0, -73.98041534423828, 40.738563537597656, -73.99948120117188, 40.73115158081055, 1.81, 16.0)) 11.05 id3858529 2 2016-01-19T11:35:24.000+0000 2016-01-19T12:10:48.000+0000 1 -73.97903 40.76394 -74.00533 40.710087 N 35.4 Tuesday 3 1 11 2.511076618143245E-7 6.39 10.83 16 0.0 List(0, 5, List(0), List(1.0)) 4.0 List(0, 6, List(4), List(1.0)) 5.0 List(0, 5, List(), List()) 11.0 List(0, 23, List(11), List(1.0)) List(0, 45, List(0, 9, 27, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, -73.9790267944336, 40.763938903808594, -74.00533294677734, 40.710086822509766, 6.39, 16.0)) 35.4 id3504673 2 2016-04-06T19:32:31.000+0000 2016-04-06T19:39:40.000+0000 1 -74.01004 40.71997 -74.01227 40.70672 N 7.15 Wednesday 4 4 19 1.359155624250055E-8 1.49 12.5 16 0.0 List(0, 5, List(0), List(1.0)) 3.0 List(0, 6, List(3), List(1.0)) 1.0 List(0, 5, List(1), List(1.0)) 1.0 List(0, 23, List(1), List(1.0)) List(0, 45, List(0, 8, 12, 17, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, 1.0, -74.01004028320312, 40.719970703125, -74.01226806640625, 40.70671844482422, 1.49, 16.0)) 7.15 id2181028 2 2016-03-26T13:30:55.000+0000 2016-03-26T13:38:10.000+0000 1 -73.97305 40.79321 -73.97292 40.78252 N 7.25 Saturday 7 3 13 8.701373594095704E-9 1.19 9.85 16 0.0 List(0, 5, List(0), List(1.0)) 1.0 List(0, 6, List(1), List(1.0)) 0.0 List(0, 5, List(0), List(1.0)) 9.0 List(0, 23, List(9), List(1.0)) List(0, 45, List(0, 6, 11, 25, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, 1.0, -73.97305297851562, 40.793209075927734, -73.9729232788086, 40.78252029418945, 1.19, 16.0)) 7.25 id0801584 2 2016-01-30T22:01:40.000+0000 2016-01-30T22:09:03.000+0000 6 -73.98286 40.742195 -73.99208 40.749184 N 7.38 Saturday 7 1 22 7.438317364467243E-9 1.1 8.94 16 4.0 List(0, 5, List(4), List(1.0)) 1.0 List(0, 6, List(1), List(1.0)) 5.0 List(0, 5, List(), List()) 4.0 List(0, 23, List(4), List(1.0)) List(0, 45, List(4, 6, 20, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, -73.98285675048828, 40.74219512939453, -73.99208068847656, 40.749183654785156, 1.1, 16.0)) 7.38 id1813257 1 2016-06-17T22:34:59.000+0000 2016-06-17T22:40:40.000+0000 4 -73.96902 40.75784 -73.957405 40.765896 N 5.68 Friday 6 6 22 1.0834136581882085E-8 1.33 14.04 16 5.0 List(0, 5, List(), List()) 0.0 List(0, 6, List(0), List(1.0)) 4.0 List(0, 5, List(4), List(1.0)) 4.0 List(0, 23, List(4), List(1.0)) List(0, 45, List(5, 15, 20, 39, 40, 41, 42, 43, 44), List(1.0, 1.0, 1.0, -73.9690170288086, 40.75783920288086, -73.95740509033203, 40.76589584350586, 1.33, 16.0)) 5.68 id1324603 2 2016-05-21T07:54:58.000+0000 2016-05-21T08:20:49.000+0000 1 -73.96928 40.79778 -73.92247 40.76056 N 25.85 Saturday 7 5 7 2.0116586779003735E-7 5.71 13.25 16 0.0 List(0, 5, List(0), List(1.0)) 1.0 List(0, 6, List(1), List(1.0)) 2.0 List(0, 5, Lis

In [93]:
splits = preppedDataDF.randomSplit([0.7, 0.3])
train_df = splits[0]
test_df = splits[1]

In [94]:
from pyspark.ml.regression import LinearRegression
niter = 500         # set by trial and error
reg = 0.0          # no regularization for now
elastic_reg = 0.0  # secondary regularization parameter (ratio of L1 to L2 regularization penalties)
tolerance = 1e-5   # set by trial and error
intercept = True   # why not indeed
lr = LinearRegression(featuresCol = 'features', labelCol='label',maxIter=niter,regParam=reg,elasticNetParam=elastic_reg,tol=tolerance)
lr_model = lr.fit(train_df)
print("Coefficients: " + str(lr_model.coefficients))
print("Intercept: " + str(lr_model.intercept))

Coefficients: [-0.623595308397,-0.210732617978,-0.503676959585,-0.134868655134,-0.467533691988,1.57977783499,0.227203711396,1.89018199202,1.71879469886,1.26140825654,-0.872973661329,0.260134793378,0.865951439823,1.24412798486,0.0945399806765,1.29759022295,7.59579822144,6.25769844411,4.94071527992,5.18563562051,5.02041617853,8.39234483358,8.4649893452,8.01532647344,8.78698206061,7.94966434753,4.6373469933,7.98141889293,7.51795414556,7.02894032007,7.5732301018,8.63072455143,4.96427038928,4.43894255071,4.05732764225,1.96912814485,3.67618578197,3.19989455287,1.92698647814,-0.472029451923,11.6112621125,-1.51031447907,-25.6026766375,2.18821304234,0.0]
Intercept: 422.5920761116984

In [95]:
from math import sqrt
def calcRMSE(df, labelCol='label'):
  """Given a DataFrame with 'label' and 'prediction' columns, calculate the root-mean-squared error
    Args:
        df: a DataFrame with labels and predictions        
    Returns:
        Float
  """
  return sqrt(float(df.rdd.map(lambda row: (row[labelCol]-row['prediction'])**2).reduce(lambda a,b: a+b)) / df.count() )

def calcRsq(df, labelCol='label'):
  """Given a DataFrame with 'label' and 'prediction' columns, calculate the R-squared diagnostic  
    Args:
        df: a DataFrame with labels and predictions        
    Returns:
        Float
  """
  npts = df.count()
  mean_label = float(df.rdd.map(lambda row: row[labelCol]).reduce(lambda a,b: a+b))/npts
  total_squared_error = (df.rdd.map(lambda row: (row[labelCol]-mean_label)**2).reduce(lambda a,b: a+b))
  squared_error = (df.rdd.map(lambda row: (row[labelCol]-row['prediction'])**2).reduce(lambda a,b: a+b))  
  return (1.0 - float(squared_error) / total_squared_error)

In [96]:
# baseline model
from pyspark.sql.functions import lit

mean_label = float(train_df.select('label').groupBy().sum().first()[0]) / train_df.count()

train_withPreds_base_df = train_df.withColumn('prediction', lit(mean_label))
test_withPreds_base_df = test_df.withColumn('prediction', lit(mean_label))

In [97]:
train_withPreds_df = lr_model.transform(train_df)
test_withPreds_df = lr_model.transform(test_df)
print('Training regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(train_withPreds_df),calcRsq(train_withPreds_df)))
print ('Training baseline \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(train_withPreds_base_df), calcRsq(train_withPreds_base_df)))
print ('')
print ('Test regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(test_withPreds_df), calcRsq(test_withPreds_df)))
print ('Testing baseline \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(test_withPreds_base_df), calcRsq(test_withPreds_base_df)))

Training regression ... RMSE: 7.780, Rsq: 0.561
Training baseline ... RMSE: 11.739, Rsq: 0.000

Test regression ... RMSE: 7.841, Rsq: 0.557
Testing baseline ... RMSE: 11.788, Rsq: -0.000

In [98]:
from pyspark.ml.regression import LinearRegression
regs = [1e-4, 1e-2, 1.0, 1.e1]
niter = 500         # set by trial and error
#reg = 1.0         # no regularization for now
elastic_reg = 0.0  # secondary regularization parameter (ratio of L1 to L2 regularization penalties)
tolerance = 1e-5   # set by trial and error
intercept = True   # why not indeed
for reg in regs:
  lr_1e_4 = LinearRegression(featuresCol = 'features', labelCol='label',maxIter=niter,regParam=reg,elasticNetParam=elastic_reg,tol=tolerance)
  lr_1e_4_model = lr_1e_4.fit(train_df)
  train_withPreds_df = lr_1e_4_model.transform(train_df)
  test_withPreds_df = lr_1e_4_model.transform(test_df)
  print(reg)
  print('Training regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(train_withPreds_df),calcRsq(train_withPreds_df)))
  print ('Test regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(test_withPreds_df), calcRsq(test_withPreds_df)))
  print('')
  

0.0001
Training regression ... RMSE: 7.657, Rsq: 0.569
Test regression ... RMSE: 8.118, Rsq: 0.540

0.01
Training regression ... RMSE: 7.657, Rsq: 0.569
Test regression ... RMSE: 8.118, Rsq: 0.540

1.0
Training regression ... RMSE: 7.718, Rsq: 0.562
Test regression ... RMSE: 8.179, Rsq: 0.533

10.0
Training regression ... RMSE: 8.689, Rsq: 0.445
Test regression ... RMSE: 9.098, Rsq: 0.422

In [99]:
from pyspark.ml.regression import GBTRegressor
gbt = GBTRegressor(featuresCol = 'features', labelCol = 'label', maxIter=10)
gbt_model = gbt.fit(train_df)
gbt_predictions_train = gbt_model.transform(train_df)
gbt_predictions_test = gbt_model.transform(test_df)

print('Training Gradient-boosted tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(gbt_predictions_train),calcRsq(gbt_predictions_train)))
print ('Test Gradient-boosted tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(gbt_predictions_test), calcRsq(gbt_predictions_test)))

Training Gradient-boosted tree regression ... RMSE: 4.320, Rsq: 0.865
Test Gradient-boosted tree regression ... RMSE: 4.208, Rsq: 0.872

In [100]:
from pyspark.ml.regression import GBTRegressor
gbt = GBTRegressor(featuresCol = 'features', labelCol = 'label', maxIter=20)
gbt_model = gbt.fit(train_df)
gbt_predictions_train = gbt_model.transform(train_df)
gbt_predictions_test = gbt_model.transform(test_df)

print('Training Gradient-boosted tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(gbt_predictions_train),calcRsq(gbt_predictions_train)))
print ('Test Gradient-boosted tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(gbt_predictions_test), calcRsq(gbt_predictions_test)))

Training Gradient-boosted tree regression ... RMSE: 3.839, Rsq: 0.893
Test Gradient-boosted tree regression ... RMSE: 3.895, Rsq: 0.890

In [101]:
from pyspark.ml.regression import GBTRegressor
maxIterList =[30,40,50]
for maxIter in maxIterList:
  gbt = GBTRegressor(featuresCol = 'features', labelCol = 'label', maxIter=maxIter)
  gbt_model = gbt.fit(train_df)
  gbt_predictions_train = gbt_model.transform(train_df)
  gbt_predictions_test = gbt_model.transform(test_df)
  
  print(maxIter)
  print('Training Gradient-boosted tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(gbt_predictions_train),calcRsq(gbt_predictions_train)))
  print ('Test Gradient-boosted tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(gbt_predictions_test), calcRsq(gbt_predictions_test)))
  print('')
  

30
Training Gradient-boosted tree regression ... RMSE: 6.973, Rsq: 0.642
Test Gradient-boosted tree regression ... RMSE: 7.489, Rsq: 0.608

40
Training Gradient-boosted tree regression ... RMSE: 6.879, Rsq: 0.652
Test Gradient-boosted tree regression ... RMSE: 7.413, Rsq: 0.616

50
Training Gradient-boosted tree regression ... RMSE: 6.815, Rsq: 0.658
Test Gradient-boosted tree regression ... RMSE: 7.368, Rsq: 0.621

In [102]:
from pyspark.ml.regression import DecisionTreeRegressor
dt = DecisionTreeRegressor(featuresCol ='features', labelCol = 'label')
dt_model = dt.fit(train_df)
dt_predictions_train = dt_model.transform(train_df)
dt_predictions_test = dt_model.transform(test_df)

print('Training Decision tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(dt_predictions_train),calcRsq(dt_predictions_train)))
print ('Test Decision tree regression \t...\t RMSE: %.3f,  Rsq: %.3f' % (calcRMSE(dt_predictions_test), calcRsq(dt_predictions_test)))

Training Decision tree regression ... RMSE: 7.632, Rsq: 0.572
Test Decision tree regression ... RMSE: 8.086, Rsq: 0.544